
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    <title>Geonomics &#8212; geonomics 0.0 documentation</title>
    <link rel="stylesheet" href="_static/alabaster.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true,
        SOURCELINK_SUFFIX: '.txt'
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
   
  <link rel="stylesheet" href="_static/custom.css" type="text/css" />
  
  <meta name="viewport" content="width=device-width, initial-scale=0.9, maximum-scale=0.9" />

  </head>
  <body>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="highlight-python"><div class="highlight"><pre><span></span>               <span class="c1">## :: ::    :::            ##</span>
         <span class="c1">##:::   :::::    :::   ::    :: :: :::##</span>
      <span class="c1">## ::::     ::           ::   ::::::::::::::##</span>
    <span class="c1">##::::::                       ::::::::: :::::: :##</span>
  <span class="c1">## :    :::                    :::    ::    :::::::::##</span>
 <span class="c1">##ggggg eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
<span class="c1">##g     e    o   o  nn  n  o   o  m   m   i   c     s    ##</span>
<span class="c1">##g     eee o     o n n n o     o mm mm   i   c     sssss##</span>
<span class="c1">##g ggg eee o     o n  nn o     o m m m   i   c         s##</span>
<span class="c1">##g   g e    o   o  n   n  o   o  m   m   i   c        ss##</span>
 <span class="c1">##gggg  eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
  <span class="c1">##  ::::::::        ::::::::::::  :       ::  ::   : ##</span>
    <span class="c1">##  ::::              :::::::  ::     ::::::::  :##</span>
      <span class="c1">## :::               :::::: ::       ::::::  ##</span>
         <span class="c1">##:                ::::                ##</span>
               <span class="c1">##                         ##</span>
</pre></div>
</div>
<div class="section" id="geonomics">
<h1>Geonomics<a class="headerlink" href="#geonomics" title="Permalink to this headline">¶</a></h1>
<div class="section" id="a-python-package-for-easy-construction-of-individual-based-spatially-explicit-forward-time-and-highly-customizable-landscape-genomics-simulations">
<h2>A Python package for easy construction of individual-based, spatially explicit, forward-time, and highly customizable landscape-genomics simulations<a class="headerlink" href="#a-python-package-for-easy-construction-of-individual-based-spatially-explicit-forward-time-and-highly-customizable-landscape-genomics-simulations" title="Permalink to this headline">¶</a></h2>
<div class="section" id="intro">
<h3>Intro<a class="headerlink" href="#intro" title="Permalink to this headline">¶</a></h3>
<p>Backward-time (i.e. coalescent) simulators abound.
But they are inadequate for simulation of many scenarios of
interest, including: natural selection on traits with arbitrary genomic
architectures; spatially variable natural selection; simulation of populations
distributed continiously and moving realistically across
complex landscapes; complex demographic change simultaneous with ongoing,
often non-stationary environmental change; and coevolutionary interactions
between multiple species or incpient species. Few existing forward-time
simulators can model all of these phenomena, and the few that can often
impose a high cost of entry (e.g. learning a new, non-standard programming
language in order to write one’s desired simulations). Geonomics aims to fill
this empty niche by combining ease of use with , all in a popular,
general-purpose scripting language. If it succeeds at doing this, Geonomics
should prove uniquely useful for a wide range of purposes, from intro-level
educational use to high-quality theoretical, methogological, empirical, and
applied research.</p>
<p>Geonomics is written in Python. Python is a full-fledged scripting language
that is relatively easy to learn (and fun!). So it can be pretty quick for a
new user to get up to speed and start doing useful work. For work with
Geonomics, this turnaround time should be even quicker. Geonomics aims to
require minimal Python knowledge (yet maintain high extensibility for
interested, expert users). Essentially, anyone should be able to build their
own, arbitrarily complex Geonomics models as long as they know how to install
the package, open a Python console, call Python functions, and edit some
default values in a pre-packaged script.</p>
<p>For proper acquaintance with the model, however, it will be helpful to
understand the concept of <strong>object-oriented programming</strong>.  Python is a very
handy language for object-oriented programming, and this is the primary
programming paradigm in which Geonomics is written. Essentially,
object-orientation involves:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Defining certain types of data structures, or <strong>classes</strong> (e.g.
<code class="code docutils literal"><span class="pre">Car</span></code>), and assigning them various behaviors, or <strong>methods</strong>
(e.g. <code class="code docutils literal"><span class="pre">honk</span></code>, <code class="code docutils literal"><span class="pre">turn</span></code>, <code class="code docutils literal"><span class="pre">brake</span></code>);</li>
<li>Using particular data values to create individual instances, or
<strong>objects</strong> belonging to those classes (e.g. <code class="code docutils literal"><span class="pre">my_1986_jeep</span></code>, or
<code class="code docutils literal"><span class="pre">batmobile</span></code>);</li>
<li>Instructing those <strong>objects</strong> to carry out their behaviors by ‘calling’
their <strong>methods</strong> (e.g. <code class="code docutils literal"><span class="pre">my_1986_jeep.honk()</span></code> might return “Beepity
beep!”, wheras <code class="code docutils literal"><span class="pre">batmobile.honk()</span></code> might
return “&lt;Batman theme song&gt;”).</li>
</ol>
</div></blockquote>
<p>Geonomics defines a number of <strong>classes</strong>, some with a number of <strong>methods</strong>
designed specifically for the user. The user will create <strong>objects</strong>
belongining to these classes, using their own particular data values
(i.e. with the values they define in the paramaters files they create). Then
the user will call key <strong>methods</strong> on these <strong>objects</strong>, in order to get them
to carry out their behaviors (that is to say, to get them to run a
simulation).</p>
<p>The subsequent documentation will present the <strong>classes</strong> definined in
Geonomics and their key <strong>methods</strong>. It will explain exactly what those methods
do, and how what they do fits into the overall structure and function of
Geonomics models.</p>
<p>In brief, Geonomics provides the following components:</p>
<blockquote>
<div><ol class="arabic simple">
<li><code class="code docutils literal"><span class="pre">Land</span></code>, <code class="code docutils literal"><span class="pre">GenomicArchitecture</span></code>, <code class="code docutils literal"><span class="pre">Trait</span></code>,
<code class="code docutils literal"><span class="pre">Individual</span></code>, <code class="code docutils literal"><span class="pre">Population</span></code>, and <code class="code docutils literal"><span class="pre">Community</span></code> classes;</li>
<li>A <code class="code docutils literal"><span class="pre">Model</span></code> class that builds objects of the aforementioned classes
according to the scenario stipulated in a parameters file,
uses them to run numerous simulation iterations, and saves data
and basic statistics from those iterations at all desired timesteps;</li>
<li>Classes for planning arbitrarily complex landscape- and
demographic-change scenarios;</li>
<li>Numerous visualization methods, to help users design, run, explore,
present, and explain their models’ behavior and results.</li>
</ol>
</div></blockquote>
<dl class="docutils">
<dt><strong>For the beginner</strong>, we recommend the following steps:</dt>
<dd><ol class="first last arabic simple">
<li>Review the following three sections (‘Model organization’, ‘Data
structures and classes’, and ‘Operations’), to get a general
undertsanding of the logic, components, and necessary and optional
behaviors of a Geonomics model;</li>
<li>Skim the subsequent section (‘Parameters’), to understand the structure
of a Geonomics parameters file;</li>
<li>Pip-install Geonomics (<code class="code bash docutils literal"><span class="pre">$</span> <span class="pre">pip</span> <span class="pre">install</span> <span class="pre">geonomics</span></code>);</li>
<li>Open Python and run <code class="code py python docutils literal"><span class="keyword namespace"><span class="pre">import</span></span> <span class="name namespace"><span class="pre">geonomics</span></span> <span class="keyword namespace"><span class="pre">as</span></span> <span class="name namespace"><span class="pre">gnx</span></span></code>;</li>
<li>Use the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> function, to begin
creating template parameters files that they can tweak as desired;</li>
<li>Use the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_model</span></span></code> function and then the
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> or <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span></code> methods to instantiate and run
the models they’ve parameterized;</li>
<li>Use the various <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot</span></span></code> methods to visualize
the behavior and results of their models.</li>
</ol>
</dd>
</dl>
<p><strong>For the `impatient` beginner</strong>, as soon as Geonomics has been
installed, you should be able to run the following code:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">geonomics</span> <span class="kn">as</span> <span class="nn">gnx</span>

<span class="n">gnx</span><span class="o">.</span><span class="n">run_default_model</span><span class="p">()</span>
</pre></div>
</div>
<p>This will load the Geonomics package, create a default Geonomics
parameters file in your current working directory,
then use that file to instantiate and run a <code class="code docutils literal"><span class="pre">Model</span></code> using the default
parameter values.</p>
<p>Lastly, please note that this documentation is designed to be read from
from the beginning (though not necessarily all the way to the end). But
given the interrelationships between all the components of the package,
you will inevitably run into material in different sections that is related.
To the extent possible, we attempt to cross-reference rather than duplicate
information.</p>
<p>Merry modeling!</p>
</div>
<hr class="docutils" />
<div class="section" id="model-organization">
<h3>Model organization<a class="headerlink" href="#model-organization" title="Permalink to this headline">¶</a></h3>
<p>&lt;&lt;&lt;DIAGRAM HERE&gt;&gt;&gt;</p>
<p>:: +——-+        +——-+
:: | stuff | —–&gt; | stuff |
:: +——-+        +——-+
::     ^                |
::     |                v
:: +——-+        +——-+
:: | stuff | &lt;—– | stuff |
:: +——-+        +——-+</p>
</div>
<hr class="docutils" />
<div class="section" id="data-structures-and-classes">
<h3>Data structures and classes<a class="headerlink" href="#data-structures-and-classes" title="Permalink to this headline">¶</a></h3>
<p>The following sections discuss the structure and function of the key
Geonomics classes. Users will interface with these classes more or less
directly when running Geonomics models, so a fundamental understanding of how
they’re organized and how they work will be useful.</p>
<div class="section" id="land-and-scape-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> objects<a class="headerlink" href="#land-and-scape-objects" title="Permalink to this headline">¶</a></h4>
<p>One of the core components of a Geonomics model is the land. The land is
modeled by the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> class. This class is an
integer-keyed <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> composed of numerous instances of the
class <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> represents a separate
environmental variable (or ‘layer’, in GIS terminology),
which is modeled a 2d Numpy array (or raster; in
attribute ‘rast’), of identical dimensions to each
other <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> in the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>
object, and with the values of its environmental variable ‘e’ constrained to
the interval [0 &lt;= e &lt;= 1]. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> can be initialized from its own
parameters subsection within the ‘land’ parameters section of a Geonomics
parameters file.</p>
<p>For each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> (see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Populations</span></span></code>’, below), the different <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>
layers in the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> can be used to model habitat
viability, habitat connectivity, or variables imposing spatially varying
natural selection. <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> objects
also contain some metatdata (as public attributes), including
the resolution (attribute ‘res’), upper-left corner (‘ulc’),
and projection (‘prj’), which default to 1, (0,0), and None but
will be set otherwise if some or all of the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> layers are read in from
real-world GIS rasters.</p>
</div>
<div class="section" id="genomes-genomicarchitecture-and-trait-objects">
<h4>Genomes, <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> objects<a class="headerlink" href="#genomes-genomicarchitecture-and-trait-objects" title="Permalink to this headline">¶</a></h4>
<p><code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> objects (see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">Individuals</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Populations</span></span></code>’, below) can optionally be assigned genomes.
If they are, each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s genome is modeled as a
2-by-L Numpy array (where 2 is the ploidy, currently fixed at
diploidy, and L is genome length) containing 0s and 1s (because
Geonomics strictly models diallelic SNPs).</p>
<p>The parameter L, as well as numerous other genomic parameters (including
locus-wise starting frequencies of the 1 alleles; locus-wise dominance effects;
locus-wise recombination rates; and genome-wide mutation rates for neutral,
globally deleterious, and adaptive loci), are controlled by the
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object pertaining to the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> to which an
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> belongs. (For the full and detailed list of attributes in a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> object, see its class documentation, below.)
The genomes of the initial <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
in a simulation, as well as those of
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s in subsequent generations, are either drawn
or recombined, and are mutated, according to the values stipulated
by the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> of
their <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>. The user can create a population with a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> and with corresponding
genomes by including a ‘genome’ subsection in that
population’s section of the Geonomics parameters file (and
setting the section’s various parameters to their desired values).</p>
<p>Geonomics can model <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ phenotypes.
It does this by allowing the
user to create an arbitrary number of distinct <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s
for each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>. Each trait is
represented by a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> object, which
maps genomic loci onto that trait, maps effect sizes (‘alpha’) onto those loci,
and sets the trait’s polygenic selection
coefficient (‘phi’). An <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s
phenotype for a given trait is calculated as the ‘null phenotype’ plus a
weighted sum of the products of its ‘effective genotypes’ at all loci
underlying that <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> and the effect sizes (i.e. ‘alpha’) of those loci:</p>
<div class="math">
\[z_{i,t} = null\_genotype + \sum_{l = 0}^{n} \alpha_{t,l} g_{i,l}\]</div>
<p>where <span class="math">\(z_{i,t}\)</span> is the phenotype of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> i for trait t,
<span class="math">\(g_{i, l}\)</span> is the genotype of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> at that locus, and
<span class="math">\(\alpha_{t,l}\)</span> is the effect size of that locus for that trait.</p>
<p>The ‘null phenotype’ refers determines what would be the phenotypic value that
an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> who is homozygyous for
the 0 allele at all loci for a trait.
For monogenic traits the null phenotype is 0 and the effect size is fixed at
0.5 (such that individuals can have phenotypes of 0, 0.5, or 1);
for polygenic traits the null phenotype is 0.5 and effect sizes can be fixed
at or distributed around a mean value (which is controlled in the
parameters file).</p>
<p>The ‘effective genotype’ refers to how the genotype is calculated based on the
dominance at a locus, as indicated by the following table of genotypes:</p>
<table border="1" class="docutils">
<colgroup>
<col width="36%" />
<col width="32%" />
<col width="32%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Biallelic genotype</th>
<th class="head">Codominant</th>
<th class="head">Dominant</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>0 : 0</td>
<td>0</td>
<td>0</td>
</tr>
<tr class="row-odd"><td>0 : 1</td>
<td>0.5</td>
<td>1</td>
</tr>
<tr class="row-even"><td>1 : 1</td>
<td>1</td>
<td>1</td>
</tr>
</tbody>
</table>
<p>(For the full and detailed list of attributes in a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> object,
see its class documentation, below.)</p>
<p>Note that for maximal control over the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>
of a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>, the user can set the value of the ‘gen_arch_file’
parameter in the parameters file to the name of a separate CSV file
stipulating the locus numbers, starting 1-allele frequencies, dominance
effects, traits, and inter-locus recombination rates (as columns) of
all loci (rows) in the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>;
these values will override any other values provided in the ‘genome’
subsection of the population’s parameters.</p>
</div>
<div class="section" id="individual-population-and-community-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> objects<a class="headerlink" href="#individual-population-and-community-objects" title="Permalink to this headline">¶</a></h4>
<p>Being that Geonomics is an individual-based model, individuals serve as
the fundamental units (or agents) of all simulations. They are represented by
objects of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> class.
Each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> has an index (saved
as attribute ‘idx’), a sex (attribute ‘sex’), an age (attribute ‘age’),
an x,y position (in continuous space; attributes ‘x’ and ‘y’), and a
<code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of environment values (attribute ‘e’), extracted from the
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s current cell on each <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> of the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> on which
the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> lives.</p>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> class is an <code class="code py python docutils literal"><span class="name"><span class="pre">OrderedDict</span></span></code>
(defined by the <code class="code py python docutils literal"><span class="name"><span class="pre">collections</span></span></code>
package) containing all <code class="code py python docutils literal"><span class="name"><span class="pre">Individaul</span></span></code>s, (with
their ‘idx’ attributes as keys). If a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>
has a <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
in the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> will also each have genomes (attribute ‘genome’),
and the <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> includes <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s
then each individual will also have a <code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> of
phenotype values (one per <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>; attribute ‘z’) and a
single fitness value (attribute ‘fit’). (These attributes all otherwise
default to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>.)</p>
<p>Each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> also has a number of other attributes of interest. Some
of these are universal (i.e. they are created regardless of the
parameterization of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> to which a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> inheres). These
include: the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s name (attribute ‘name’); its current density
raster (a Numpy array attribute called ‘N’); and the number of births,
number of deaths, and terminal population size of each timestep (which are
<code class="code py python docutils literal"><span class="name builtin"><span class="pre">list</span></span></code> attributes called ‘n_births’, ‘n_deaths’, and ‘Nt’). If the
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> was parameterized with a
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code> then that will
be created as the ‘gen_arch’ attribute (otherwise this attribute will be
<code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">None</span></span></code>).</p>
<p>All of the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s in a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>
are collected in the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>’s
<code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> object. The <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> class
is simply an integer-keyed <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code>
of <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s. For the time being, the <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> object allows a
Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> to simulate multiple <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s simultaneously on
the same <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>, but otherwise affords no additional functionality
of interest. However, its implementation will facilitate the potential
future development of methods for interaction between <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s.
(e.g. to simulate coevolutionary, speciation, or hybridization scenarios).</p>
</div>
<div class="section" id="model-objects">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> Objects<a class="headerlink" href="#model-objects" title="Permalink to this headline">¶</a></h4>
<p>Objects of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> class serve as the main interface between the user
and the Geonomics program. (While it is certainly possible for a user
to work directly with the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>
and <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> or <code class="code py python docutils literal"><span class="name"><span class="pre">Community</span></span></code> objects to
script their own custom models, the typical user should find that the
<code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object allows them accomplish their goals with minimal toil.)
The main affordance of a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object is the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span></code> method,
which, as one could guess, will run the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>. The typical workflow
for creating and running a  <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object is as follows:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Create a template paramters file containing the desired sections,
by calling <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> with all revelant arguments;</li>
<li>Define the scenario to be simulated, by opening and editing that
parameters file (and optionally, creating/editing corresponding
files, e.g. genomic-architecture CSV files;
or raster or numpy-array files to be used as <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s);</li>
<li>Instantiate a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object from that parameters file, by calling
<code class="code py python docutils literal"><span class="name"><span class="pre">mod</span></span> <span class="operator"><span class="pre">=</span></span> <span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_model</span></span><span class="punctuation"><span class="pre">(</span></span><span class="literal string single"><span class="pre">‘/path/to/params_filename.py’</span></span><span class="punctuation"><span class="pre">)</span></span></code>;</li>
<li>Run the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, by calling <code class="code py python docutils literal"><span class="name"><span class="pre">mod</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">run</span></span><span class="punctuation"><span class="pre">()</span></span></code>.</li>
</ol>
</div></blockquote>
<p>For detailed information on usage of these functions, see their docstrings.
When a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is run, it will:</p>
<blockquote>
<div><ol class="arabic simple">
<li>Run the burn-in (until the mininmal burn-in length stipulated in the
parameters file and the built-in stationarity statistics
determine that the burn-in is complete);</li>
<li>Run the main model for the stipulated number of timesteps;</li>
<li>Repeat this for the stipulated number of iterations (retaining or
refreshing the first run’s initial <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>
objects as stipulated).</li>
</ol>
</div></blockquote>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object offers one other method, however, <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code>,
which allows the user to run a model, in either ‘burn’ or ‘main’ mode,
for an arbitrary number of timesteps within a single iteration (see its
docstring for details). This is particularly useful for running
Geonomics within an interactive Python session. Thus, <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is
primarily designed for passively running numerous iterations of a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>,
to generate data for analysis, whereas <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">walk</span></span></code> is primarily designed
for the purposes of learning, teaching, or debugging the package, or
developing, exploring, introspecting, or visaulizing particular <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>s.</p>
</div>
<div class="section" id="secondary-classes">
<h4>Secondary classes<a class="headerlink" href="#secondary-classes" title="Permalink to this headline">¶</a></h4>
<p>The typical user will not need to access or interact with the following
classes in any way. They will, however, parameterize them in the
parameters file by either leaving or altering their default values. Geonomics
sets generally sensible default parameter values wherever possible,
but for some scenarios they may not be adequate, and for some parameters
(e.g. the window-width used by the _DensityGridStack; see below), there is
no “one-size-fits-most” option. Thus, it is important that the user
have a basic acquaintance with the purpose and operation of these classes.</p>
<div class="section" id="movementsurface">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code><a class="headerlink" href="#movementsurface" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> class allows Geonomics
to model a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s
realistic movement across a spatially varying landscape. It does this by
creating an array of circular probability distributions (i.e. VonMises
distributions), one for each cell on the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>, from which
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s choose their directions each time they move. To create the
<code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> for a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>,
the user must indicate the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>
that should be used to create it (i.e. the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> that represents
landscape permeability for that <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>). The <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code>’s
distributions can be <strong>simple (i.e. unimodal)</strong>, such that the
maximum value of the distribution at each cell will point toward the
maximum value in the 8-cell neighborhood; this works best for permeability
<code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s with shallow, monotonic gradients, because the differences
between permeability values of neighboring cells can be minor (e.g. a
gradient representing the directionality of a prevalent current).
Alternatively, the distributions can be <strong>mixture (i.e. multimodal)</strong>
distributions, which are weighted sums of 8 unimodal distributions, one
for each neighboring cell, where the weights are the relative cell
permeabilities (i.e. the relative probabilities that an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> would
move into each of the 8 neighboring cells); this works best for non-monotonic,
complex permeability <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s (e.g. a DEM of a mountainous region that is
used as a permeability <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>). (The <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> is surrounded by a margin
of 0-permeability cells before the <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> is calculated, such
that <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> edges are treated as barriers to movement.) The class consists
principally of a 3d Numpy array (x by y by z, where x and y are the
dimensions of the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> and z is the length of the vector of values
used to approximate the distributions in each cell.</p>
</div>
<div class="section" id="densitygridstack">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_DensityGridStack</span></span></code><a class="headerlink" href="#densitygridstack" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_DensityGridStack</span></span></code> class implements an algorithm for rapid estimating
an array of the local density of a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>. The density is estimated
using a sliding window approach, with the window-width determining the
neighborhood size of the estimate. The resulting array has a spatial
resolution equivalent to that of the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>, and is used in all
density-dependent operations.</p>
</div>
<div class="section" id="kdtree">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_KDTree</span></span></code><a class="headerlink" href="#kdtree" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_KDTree</span></span></code> class is just a wrapper around <code class="code py python docutils literal"><span class="name"><span class="pre">scipy</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">spatial</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">cKDTree</span></span></code>.
It provides an optimized algorithm (the kd-tree) for finding
neighboring points within a given search radius.
This class is used for all neighbor-searching operations (e.g. mate-search).</p>
</div>
<div class="section" id="recombinationpaths">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code><a class="headerlink" href="#recombinationpaths" title="Permalink to this headline">¶</a></h5>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code> class contains a large (and customizable)
number of <code class="code py python docutils literal"><span class="name"><span class="pre">bitarray</span></span></code>s, each of which indicates the genome-length
diploid chromatid numbers (0 or 1) for a
recombinant gamete produced by an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> of a given <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>
(henceforth referred to as ‘recombination paths’). These recombination
paths are generated using the genome-wide recombination rates specified by
the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s <code class="code py python docutils literal"><span class="name"><span class="pre">GeonomicArchitecture</span></span></code>. They are generated during
construction of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, then drawn randomly as needed (i.e.
each time an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> produces a gamete). This provides a
reasonable trade-off between realistic modelling of recombination and runtime.</p>
</div>
<div class="section" id="landchanger-and-populationchanger">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_LandChanger</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">_PopulationChanger</span></span></code><a class="headerlink" href="#landchanger-and-populationchanger" title="Permalink to this headline">¶</a></h5>
<p>These classes manage all of the landscape changes and demographic changes
that were parameterized for the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> and
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> objects to which they inhere.
The functions creating these changes are defined at the outset,
then queued and called at their scheduled timesteps.</p>
</div>
<div class="section" id="datacollector-and-statscollector">
<h5><code class="code py python docutils literal"><span class="name"><span class="pre">_DataCollector</span></span></code> and <code class="code py python docutils literal"><span class="name"><span class="pre">_StatsCollector</span></span></code><a class="headerlink" href="#datacollector-and-statscollector" title="Permalink to this headline">¶</a></h5>
<p>These classes manage all of the data and statistics that should be collected
and written to file for the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object to which they inhere
(as determined by the parameters file used the create the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>).
The types of data to be collected, or statistics to be calculated, as
well as the timesteps at which and methods by which they’re
collected/calculated and determined at the outset, then the
appropriate functions called at the appropriate timesteps.</p>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="operations">
<h3>Operations<a class="headerlink" href="#operations" title="Permalink to this headline">¶</a></h3>
<p>The following sections discuss the mechanics of core Geonomics operations.
The material here is inevitably intertwined with some of the material in
the “Data structures and classes” section. To the extent possible, we
attempt to cross-reference rather than duplicate information (with
the exception of this sentence).</p>
<div class="section" id="movement-and-dispersal">
<h4>Movement and Dispersal<a class="headerlink" href="#movement-and-dispersal" title="Permalink to this headline">¶</a></h4>
<p>Movement is optional, such that turning off movement will allow the user
to simulate sessile organisms (which will reproduce and disperse,
but not move after dispersal; this distinction is of course irrelevant
for a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> with a maximum age of 1). For <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s
with movement, <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s can
move by two distinct mechanisms. <strong>Spatially random movement</strong>
is the default behavior; in this case, <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
move to next locations that are determined by a random distance drawn
from a Wald distribution and a random direction drawn from a uniform
circular (i.e. Von Mises) distribution.  As with most distributions used
in Geonomics, the parameters of these distributions have sensible
default values but can be customized in a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>’s parameters file
(see section ‘Parameters’, below).</p>
<p>The alternative movement mechanism that is available is
<strong>movement across a permeability surface</strong>,
using a <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> object.
To parameterize a <code class="code py python docutils literal"><span class="name"><span class="pre">_MovemementSurface</span></span></code> for a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>, the user
must create a template parameters file that includes the
necessary parameters section for the population (i.e.
the user must set ‘movement’ to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">True</span></span></code> and ‘movement_surface’ to <code class="code py python docutils literal"><span class="name builtin pseudo"><span class="pre">True</span></span></code>
in the population’s arguments to the <code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code>
function (see the docstring for that function for details and an example).
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s move to next locations determined by a random distance drawn
from a Wald distribution and a random direction drawn from the distribution
at the  <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> cell in which which the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s
are currently located. For details about <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> creation, see
section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code>’ above, or the class’ docstring.</p>
<p>Dispersal is currently implemeneted identically to spatially random movement
(with the caveat that the an offspring’s new location is determined
relative its parents’ centroid). But the option to use a
<code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> for dispersal will be offered soon.</p>
</div>
<div class="section" id="reproduction">
<h4>Reproduction<a class="headerlink" href="#reproduction" title="Permalink to this headline">¶</a></h4>
<p>Each timestep, for each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>, all pairs of individuals within
a certain distance of each other (i.e. the mating radius,
which is set in the parameters file) are identified.
These pairs are subsetted if necessary (i.e. if the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>
requires that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s be above a certain reproductive age,
or that they be of opposite sexes, in order to mate; these values
can also be changed from their defaults in the parameters file).
Remaining pairs mate probabilistically (according to a Bernoulli
random draw with probability equal to the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s birth
rate, which is also set in the parameters file).</p>
<p>Pairs that are chosen to mate will produce a number of new
offspring drawn from a Poisson distribution (with lambda set in the
parameters file). For each offspring, sex is chosen probablistically
(a Bernoulli random draw with probability equal to the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s
sex ratio), age set to 0, and location chosen by dispersal from
the parents’ centroid (see section ‘Movement and Dispersal’). For
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s that have genomes, offspring genomes will be a
fusion of two recombinant genomes from each of the two parents (where
each recombinant is indexed out a parent’s genome using a recombination
path; see section ‘<code class="code py python docutils literal"><span class="name"><span class="pre">_RecombinationPaths</span></span></code>’). For <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s
with <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s in their
<code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>s, offspring phenotypes are
determined at birth. Mutations are also drawn and introduced at this
point (see section ‘Mutation for details).</p>
</div>
<div class="section" id="mortality">
<h4>Mortality<a class="headerlink" href="#mortality" title="Permalink to this headline">¶</a></h4>
<p>Mortality can occur as a combination of two factors: <strong>density dependence</strong>
and <strong>natural selection</strong>. Each <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> has a death decision drawn
as a Bernoulli random variable with
<span class="math">\(P(d_{i}) = 1 - P(s_{i_{dens}})P(s_{i_{fit}})\)</span>, where <span class="math">\(P(d_{i})\)</span>
is the probability of death of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>, and
<span class="math">\(P(s_{i_{dens}})\)</span> and <span class="math">\(P(s_{i_{fit}})\)</span> are the probabilities of
survival of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span> given density-dependence and
fitness. The probability of density-dependent death is contingent on an
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s x,y location
(i.e. the cell in which they’re currently located.
And an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s probability of survival due to fitness
is just equal to the product of their absolute fitness (<span class="math">\(\omega\)</span>)
for each of the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s <span class="math">\(m\)</span> <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>s.
Thus the equation for an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s probability of death becomes:</p>
<div class="math">
\[P(d_{i}) = 1 - (1 - P(d_{x,y})) \prod_{p = 1}^{m}\omega_{i,p}\]</div>
<p>The following two sections explain in detail the implementation and
calculation of the two halves of the right side of this equation.</p>
<div class="section" id="density-dependence">
<h5>Density dependence<a class="headerlink" href="#density-dependence" title="Permalink to this headline">¶</a></h5>
<p>Density dependence is implemented using a spatialized form of the class
logistic growth equation
(<span class="math">\(\frac{\mathrm{d}N_{x,y}}{\mathrm{d}t}=rN_{x,y}(1-\frac{N_{x,y}}{K_{x,y}})\)</span>,
where the x,y subscripts refer to values for a given cell on the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>).
Each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> has a carrying-capacity raster (a 2d Numpy array;
attribute ‘K’), which is defined in the parameters file to be
one of the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s in the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>.
The comparison between this raster and
the population-density raster calculated at each timestep serves as the
basis for the spatialized logistic growth equation, because both
equations can be calculated cell-wise for the entire extent of the
<code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> (using the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>’s intrinsic growth rate, the attribute
‘R’, which is set in the parameters file).</p>
<p>The logistic equation returns an array of instantaneous population growth
rates within each cell. We can derive from this the density-dependent
probability of death at each cell by subtracting an array of the expected
number of births at each cell, then dividing by the array of
population density:</p>
<div class="math">
\[P(d_{x,y}) = E[N_{d;x,y}]/N_{x,y} = \frac{E[N_{b;x,y}]
 - \frac{\mathrm{d}N_{x,y}}{\mathrm{d}t}}{N_{x,y}}\]</div>
<p>The expected number of births at each cell is calculated as a density
raster of the number of succesful mating pairs, multiplied by the expected
number of births per pair (i.e. the expectation of the Poisson
distribution of the number of offspring per mating pair, which
is just the distribution’s paramater lambda).</p>
</div>
<div class="section" id="selection">
<h5>Selection<a class="headerlink" href="#selection" title="Permalink to this headline">¶</a></h5>
<p>Selection on a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> can exhibit three regimes: <strong>spatially divergent</strong>,
<strong>universal</strong>, and <strong>spatially contingent</strong>. <strong>Spatially divergent</strong> selection
is the default behavior, and the most commonly used; in this form of
selection, an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s fitness depends on the absolute difference
between the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s phenotypic value and the environmental
value of the relevant <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> (i.e. the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> that represents the
environmental variable acting as the selective force) in the cell where
the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> is located.</p>
<p><strong>Universal</strong> selection (which can be toggled using the ‘univ_adv’
parameter with a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s section in the parameters file) occurs
when a phenotype of 1 is optimal everywhere on the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>. In other
words, it represents directional selection on an entire <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>,
regardless of <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>s’ spatial contexts. (Note that this can
be thought of as operating the same as spatially divergent selection,
but with the environmental variable driving natural selection being
represented by an array in which all cells are equal to 1.)</p>
<p>Under <strong>spatially contingent</strong> selection, the selection coefficient of a
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> varies across space, such that the strength of selection
is environmentally determined in some way. Importantly, this selection regime
is <em>not mutually exclusive</em> with the other two; in other words,
selection on a certain <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> be both spatially contingent
and either spatially divergent or universal. Spatially contingent selection
can be implemented by providing an array of values (equal in dimensions
to the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>) to the ‘phi’ value of a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>, rather than a scalar
value (which could be done within the parameters file itself, but may be
more easily accomplished as a step between reading in a parameters file and
instantiating a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> object from it). (Note that non-spatailly
cotingent selection could in fact be thought of as a special case of
spatially contingent selection, but where the array of selection-coefficients
has the same value at each cell.)</p>
<p>All possible combinations of the three selection regimes of selection can all
be thought of as special cases of the following equation for the fitness of
<code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span> for <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> <span class="math">\(p\)</span> (<span class="math">\(\\omega_{i,p}\)</span>):</p>
<div class="math">
\[\omega_{i,p}= 1 - \phi_{p;x,y} (\mid e_{p;x,y} - z_{i;p} \mid)^{\gamma_{p}}\]</div>
<p>Where <span class="math">\(\\phi_{p;x,y}\)</span> is the selection coefficient of trait
<span class="math">\(p\)</span>, <span class="math">\(e_{p;x,y}\)</span> is the environmental variable of the
relevant <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> at <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>‘s x,y location
(which can also be thought of as the <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s optimal
phenotype), <span class="math">\(z_{i;p}\)</span> is <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> <span class="math">\(i\)</span>‘s (actual)
phenotype for <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> <span class="math">\(p\)</span>, and <span class="math">\(gamma_{p}\)</span> controls
how fitness decreases as the absolute difference between an <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s
optimal and actual phenotypes increases (it defaults to 1, which causes
fitness to decrease linearly around the optimal
phenotypic value).</p>
</div>
</div>
<div class="section" id="mutation">
<h4>Mutation<a class="headerlink" href="#mutation" title="Permalink to this headline">¶</a></h4>
<p>Geonomics can model mutations of three different types: <strong>neutral</strong>,
<strong>deleterious</strong>, and <strong>trait</strong> mutations. These terms don’t map
precisely onto the traditional population-genetic
lingo of “neutral”, “deleterious”, and “beneficial”, but they
are more or less analogous:</p>
<ul class="simple">
<li><strong>Neutral</strong> mutations are the same conceptually in Geonomics as
they are in the field of population genetics in general:
They are mutations that have no effect on the fitness of
the individuals in which they occur.</li>
<li><strong>Deleterious</strong> mutations in Geonomics are also conceptually the
same in Geonomics and in population genetics: They negatively impact
the fitness of the individuals in which they occur.</li>
<li><strong>Trait</strong> mutations are the place where the Geonomics concept and
the population-genetic concept diverge: In Geonomics, natural selection
acts on the phenotype, not the genotype (although these concepts are
identical if a <code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code> in monogenic), and it is (by default,
but not always; see section ‘Selection’, above) divergent. For this reason
it would be a misnomer to call mutations that influence a given
<code class="code py python docutils literal"><span class="name"><span class="pre">Trait</span></span></code>’s phenotypes ‘beneficial’ – even though that term is the closest
population-genetic concept to this concept as it is employed in Geonomics –
because the same mutant genotype in the same <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code> could have opposite
effects on that <code class="code py python docutils literal"><span class="name"><span class="pre">Individual</span></span></code>’s fitness in different environmental
contexts (i.e. it could behave as a beneficial mutation is one region of
the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> but as a deleterious mutation in another).</li>
</ul>
</div>
<div class="section" id="population-interaction">
<h4>Population interaction<a class="headerlink" href="#population-interaction" title="Permalink to this headline">¶</a></h4>
<p>This functionality is not yet included available. But the Community class was
created in advance recognition that this functionality could be desirable
for future versions (e.g. to simulate coevolutionary, speciation, or
hybridization scenarios).</p>
</div>
<div class="section" id="land-and-population-change">
<h4>Land and population change<a class="headerlink" href="#land-and-population-change" title="Permalink to this headline">¶</a></h4>
<p>For a given <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>, any number of landsacpe change events can be planned.
In the parameters file, for each event, the user stipulates the initial
timestep; the final timestep; the resulting landscape (i.e. the array
of the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> that will exist after the event is complete); and the
interval at which intermediate changes will occur.  When the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is
created, the stepped series of intermediate landscapes (and
<code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> objects, if the <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code> that is changing serves
as the basis of a <code class="code py python docutils literal"><span class="name"><span class="pre">_MovementSurface</span></span></code> for any <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>) will be
created and queued, so that they will swap out accordingly at the appropriate
timesteps.</p>
<p>For a given <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>, any number of demographic change events can
also be planned. In the parameters file, for each event, the user
stipulates the type of the event (‘monotonic’, ‘cyclical’, ‘random’, or
‘custom’) as well as the values of a number of associated
parameters (precisely which parameters depdends on the type of event chosen).
As with landscape change events, all necessary stepwise changes will be
planned and queued when the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is created, and will be
executed at the appropriate timesteps.</p>
<p>It is also possible to schedule any number of instantaneous changes
to some of the life-history parameters of a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> (e.g. birth rate;
the lambda parameter of the Poisson distribution determining the number of
offspring of mating events). This functionality is currently minimalistic,
but will be more facilitated in future versions.</p>
</div>
</div>
<hr class="docutils" />
<div class="section" id="visualization">
<h3>Visualization<a class="headerlink" href="#visualization" title="Permalink to this headline">¶</a></h3>
<p>Each <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> has a wide variety of visualization methods
(<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot</span></span></code>, <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_fitness</span></span></code>, etc.),
which aim to help users design, run, explore, present,
and explain their models’ behavior and results.
These methods can be called on a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> at any time (e.g. as
soon as the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> has been created, or after the model has
run for any number of timesteps); but it is worth mentioning that some
methods may be invalid depending on the point in model-time at
which they’re called (e.g.  <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_genotype</span></span></code>,
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_phenotype</span></span></code>, and <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_fitness</span></span></code>
cannot be run for Populations that have not yet been burned in,
as they will not yet have genomes assigned) or
the <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> on which they’re called
(e.g. the aforementioned methods cannot create plots for a <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>
that has no <code class="code py python docutils literal"><span class="name"><span class="pre">GenomicArchitecture</span></span></code>; and likewise, the
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">plot_demographic_changes</span></span></code> method cannot be called for a
<code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code> for which demographic changes were not parameterized).</p>
<p>The <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> object and its <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s also both have a <code class="code py python docutils literal"><span class="name"><span class="pre">plot</span></span></code> method.</p>
</div>
<hr class="docutils" />
<div class="section" id="parameters">
<h3>Parameters<a class="headerlink" href="#parameters" title="Permalink to this headline">¶</a></h3>
<p>In order to create and run a Geonomics <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>, you will need a valid
Geonomics parameters file. No worry though – this is very easy to create!
To generate a new, template parameters file, you will simply call the
<code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code> function, feeding it the appropriate
arguments (to indicate how many <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s and <code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s you
want to include in your <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>; which parameters sections you want
included in the file, both for those
<code class="code py python docutils literal"><span class="name"><span class="pre">Scape</span></span></code>s and <code class="code py python docutils literal"><span class="name"><span class="pre">Population</span></span></code>s and for
other components of the <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code>; and the path and filename for your new
parameters file). Geonomics will then automatically create the file for you,
arranged as you requested and saved where you requested.</p>
<p>When you then open that file, you will see the following:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#&lt;your_filename&gt;.py</span>

<span class="c1">#This is a default parameters file generated by Geonomics</span>
<span class="c1">#(by the gnx.params.make_parameters_file() function).</span>


                      <span class="c1">## :: ::    :::            ##</span>
                <span class="c1">##:::   :::::    :::   ::    :: :: :::##</span>
             <span class="c1">## ::::     ::           ::   ::::::::::::::##</span>
           <span class="c1">##::::::                       ::::::::: :::::: :##</span>
         <span class="c1">## :    :::                    :::    ::    :::::::::##</span>
        <span class="c1">##ggggg eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
       <span class="c1">##g     e    o   o  nn  n  o   o  m   m   i   c     s    ##</span>
       <span class="c1">##g     eee o     o n n n o     o mm mm   i   c     sssss##</span>
       <span class="c1">##g ggg eee o     o n  nn o     o m m m   i   c         s##</span>
       <span class="c1">##g   g e    o   o  n   n  o   o  m   m   i   c        ss##</span>
        <span class="c1">##gggg  eeee ooo   n   n   ooo   m   m iiiii  cccc ssss##</span>
         <span class="c1">##  ::::::::        ::::::::::::  :       ::  ::   : ##</span>
           <span class="c1">##  ::::              :::::::  ::     ::::::::  :##</span>
             <span class="c1">## :::               :::::: ::       ::::::  ##</span>
                <span class="c1">##:                ::::                ##</span>
                      <span class="c1">##                         ##</span>


<span class="n">params</span> <span class="o">=</span> <span class="p">{</span>

<span class="c1">##############</span>
<span class="c1">#### LAND ####</span>
<span class="c1">##############</span>
    <span class="s1">&#39;land&#39;</span><span class="p">:</span> <span class="p">{</span>

    <span class="c1">##############</span>
    <span class="c1">#### main ####</span>
    <span class="c1">##############</span>
        <span class="s1">&#39;main&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1"># dimensions of the Land</span>
            <span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>

   <span class="c1">#.</span>
   <span class="c1">#.</span>
   <span class="c1">#.</span>
</pre></div>
</div>
<p>This is the beginning of a file that is really just a long but simple Python
script (hence the ‘.py’ extension); this whole file just defines a single,
long, nested <code class="code py python docutils literal"><span class="name builtin"><span class="pre">dict</span></span></code> (i.e. a Python ‘dictionary’) containing all of your
parameter values. It may look like a lot, but don’t be concerned! For two
reasons:</p>
<blockquote>
<div><ol class="arabic simple">
<li>All the hard work is already done for you. You’ll just need to change
the default values where and how you want to, to set up your particular
simulation scenario.</li>
<li>You will probably leave a good number of the parameters defined in this
file untouched. Geonomics does its best to set sensible default values
for all its parameters. Though of course, you’ll want to think clearly
nonetheless about whether the default value for each parameter
is satisfactory for your purposes.</li>
</ol>
</div></blockquote>
<p>Each parameter in the parameters value is preceded by a terse comment, to
remind you what the parameter does. But for detailed information about each
parameter, you’ll want to refer to the following information.
What follows is a list of all of the Geonomics parameters (in the sections and
the top-to-bottom order in which they’ll appear in your parameters files).
For each parameter, you will see a snippet of the context (i.e. lines of
Python code) in which it appears in a parameters filewith; that will be
followed by detailed information about the parameter (the valid Python data
types it can take; its default values; then an explanation of what it
defines, how its value is used, and any other relevant information). This
will look like:</p>
<p><strong>&lt;param_name&gt;</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1">#brief comment about the parameter</span>
<span class="s1">&#39;&lt;param_name&gt;&#39;</span><span class="p">:</span>               <span class="o">&lt;</span><span class="n">default_param_value</span><span class="o">&gt;</span><span class="p">,</span>
</pre></div>
</div>
<p>&lt;valid Python data type&gt;</p>
<p>&lt;default value&gt;</p>
<blockquote>
<div>&lt;Explanation of what the parameter defines, how its value is used,
and any other relevant information.&gt;</div></blockquote>
<p>This section should serve as your primary point of reference
if you confront any uncertainty while creating your own parameters files.
We’ll start with the section of parameters that
pertains to the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> object.</p>
<div class="section" id="land-parameters">
<h4>Land parameters<a class="headerlink" href="#land-parameters" title="Permalink to this headline">¶</a></h4>
<p><strong>dim</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># dimensions of the Land</span>
<span class="s1">&#39;dim&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">),</span>
</pre></div>
</div>
<p>tuple;</p>
<p>default: (20,20)</p>
<blockquote>
<div>This defines the x and y dimensions of the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code>, in units of cells. As
you might imagine, these values are used for a wide variety of basic
operations throughout Geonomics. You will want to change the
default value to the dimensions of the landscape you wish to simulate on.</div></blockquote>
<p><strong>res</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># resolution of the Land</span>
<span class="s1">&#39;res&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">1</span><span class="p">),</span>
</pre></div>
</div>
<p>tuple</p>
<p>default: (1,1)</p>
<blockquote>
<div>This defines the <code class="code py python docutils literal"><span class="name"><span class="pre">Land</span></span></code> resolution (or cell-size) in the x and y
dimensions. These sizes are usually not used when a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is run, but
can be used for crosswalking <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> content with
real-world data. Defaults to the meaningless value (1,1), and this value
generally needn’t be changed in your parameters file, but it will
be automatically updated to the resolution of any GIS rasters that
are read in for use as <code class="code py python docutils literal"><span class="name"><span class="pre">Scapes</span></span></code>.</div></blockquote>
<p><strong>ulc</strong></p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="c1"># upper-left corner of the Land</span>
<span class="s1">&#39;ulc&#39;</span><span class="p">:</span>                      <span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">),</span>
</pre></div>
</div>
<p>tuple</p>
<p>default: (0,0)</p>
<blockquote>
<div>This defines the upper-left corner (ULC) of the landscape (in the units of
some real-world coordinate reference system, e.g. decimal degrees, or
meters). This information is not used when a <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> is run, but can be
used for crosswalking <code class="code py python docutils literal"><span class="name"><span class="pre">Model</span></span></code> content with real-world data.
Defaults to the meaningless value (0,0), and this value usually needn’t
be changed in your parameters file, but it will be automatically
updated to match the ULC value of any GIS rasters
that are read in for use as <code class="code py python docutils literal"><span class="name"><span class="pre">Scapes</span></span></code>.</div></blockquote>
<div class="highlight-python"><div class="highlight"><pre><span></span>            <span class="s1">&#39;prj&#39;</span><span class="p">:</span>                      <span class="bp">None</span><span class="p">,</span>
                <span class="c1">#projection of the landscape; only applicable if layers are to</span>
                <span class="c1">#be read in from a raster file; defaults to None</span>

    <span class="c1">################</span>
    <span class="c1">#### scapes ####</span>
    <span class="c1">################</span>
        <span class="s1">&#39;scapes&#39;</span><span class="p">:</span> <span class="p">{</span>

            <span class="mi">0</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1">#scape name; each scape must be give a unique string or numeric</span>
                <span class="c1">#(e.g. 0, 0.1, &#39;scape0&#39;, &#39;1994&#39;, &#39;mean_T&#39;); default to serial</span>
                <span class="c1">#integers from 0</span>

        <span class="c1">############################</span>
        <span class="c1">#### scape num. 0: init ####</span>
        <span class="c1">############################</span>

                <span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="c1">#initiating parameters for this scape</span>

                    <span class="s1">&#39;rand&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#parameters for making a random scape (interpolated</span>
                        <span class="c1">#from randomly located random values)</span>
                        <span class="s1">&#39;n_pts&#39;</span><span class="p">:</span>                        <span class="mi">500</span><span class="p">,</span>
                            <span class="c1">#number of random coordinates to be used in generating random landscapes</span>
                                <span class="c1">#(only needed if rand == True)</span>
                        <span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="s1">&#39;cubic&#39;</span>
                            <span class="c1"># interpolation method (valid: &#39;linear&#39;, &#39;cubic&#39;, &#39;nearest&#39;)</span>
                        <span class="p">},</span>

                    <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;init&#39;</span>

        <span class="c1">##############################</span>
        <span class="c1">#### scape num. 0: change ####</span>
        <span class="c1">##############################</span>

                <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="c1">#land-change events for this scape</span>
                    <span class="s1">&#39;end_rast&#39;</span><span class="p">:</span>         <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="mi">20</span><span class="p">,</span><span class="mi">20</span><span class="p">)),</span>
                        <span class="c1">#scape to be set as the endpoint of the land-change event</span>
                    <span class="s1">&#39;start_t&#39;</span><span class="p">:</span>          <span class="mi">1500</span><span class="p">,</span>
                        <span class="c1">#timestep on which to start the land-change event</span>
                    <span class="s1">&#39;end_t&#39;</span><span class="p">:</span>            <span class="mi">2000</span><span class="p">,</span>
                        <span class="c1">#timestep on which to end the land-change event</span>
                    <span class="s1">&#39;n_steps&#39;</span><span class="p">:</span>          <span class="mi">10</span>
                        <span class="c1">#number of stepwise changes to make between t_start and t_end</span>
                        <span class="p">},</span>

                <span class="p">},</span> <span class="c1"># &lt;END&gt; scape num. 0</span>


            <span class="mi">1</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1">#scape name; each scape must be give a unique string or numeric</span>
                <span class="c1">#(e.g. 0, 0.1, &#39;scape0&#39;, &#39;1994&#39;, &#39;mean_T&#39;); default to serial</span>
                <span class="c1">#integers from 0</span>

        <span class="c1">############################</span>
        <span class="c1">#### scape num. 1: init ####</span>
        <span class="c1">############################</span>

                <span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="c1">#initiating parameters for this scape</span>

                    <span class="s1">&#39;nlmpy&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#parameters for making an nlmpy scape (using any nlmpy</span>
                        <span class="c1">#function for which valid arguments are provided)</span>
                        <span class="c1">#NOTE: all other parameters in this dictionary will be</span>
                        <span class="c1">#unpacked as arguments to the nlmpy function chosen;</span>
                        <span class="c1">#thus, all necessary and only valid arguments must be</span>
                        <span class="c1">#provided or a general error will be thrown</span>
                        <span class="s1">&#39;function&#39;</span><span class="p">:</span>                 <span class="s1">&#39;mpd&#39;</span><span class="p">,</span>
                            <span class="c1">#filepath to read into this scape</span>
                        <span class="s1">&#39;nRow&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
                            <span class="c1">#number of rows (must agree with land dimensions)</span>
                        <span class="s1">&#39;nCol&#39;</span><span class="p">:</span>                     <span class="mi">20</span><span class="p">,</span>
                            <span class="c1">#number of columns (must agree with land</span>
                            <span class="c1">#dimensions)</span>
                        <span class="s1">&#39;h&#39;</span><span class="p">:</span>                        <span class="mi">1</span><span class="p">,</span>
                            <span class="c1">#&quot;controls the level of spatial autocorrelation</span>
                            <span class="c1">#in element values&quot;</span>
                        <span class="p">},</span>

                    <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;init&#39;</span>

                <span class="p">},</span> <span class="c1"># &lt;END&gt; scape num. 1</span>


            <span class="mi">2</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1">#scape name; each scape must be give a unique string or numeric</span>
                <span class="c1">#(e.g. 0, 0.1, &#39;scape0&#39;, &#39;1994&#39;, &#39;mean_T&#39;); default to serial</span>
                <span class="c1">#integers from 0</span>

        <span class="c1">############################</span>
        <span class="c1">#### scape num. 2: init ####</span>
        <span class="c1">############################</span>

                <span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="c1">#initiating parameters for this scape</span>

                    <span class="s1">&#39;defined&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#parameters for making a defined scape (interpolated</span>
                        <span class="c1">#from a provided set of valued points)</span>
                        <span class="s1">&#39;pts&#39;</span><span class="p">:</span>                    <span class="bp">None</span><span class="p">,</span>
                            <span class="c1">#coords of points to use to interpolate defined scape (provided as</span>
                            <span class="c1">#a nx2 Numpy array, where n matches the number of points in</span>
                            <span class="c1">#the scape_pt_vals array, to be used as the points</span>
                            <span class="c1">#to be interpolated; only needed if rand == False)</span>
                        <span class="s1">&#39;vals&#39;</span><span class="p">:</span>                      <span class="bp">None</span><span class="p">,</span>
                            <span class="c1">#point values to use to interpolate defined landscape layers (a 1xn Numpy array,</span>
                                <span class="c1">#where n matches the number of points in scape_pt_coords arrays;</span>
                                <span class="c1">#only needed if rand == False)</span>
                        <span class="s1">&#39;interp_method&#39;</span><span class="p">:</span>                <span class="bp">None</span>
                            <span class="c1"># interpolation method (valid: &#39;linear&#39;, &#39;cubic&#39;, &#39;nearest&#39;)</span>
                        <span class="p">},</span>

                    <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;init&#39;</span>

                <span class="p">},</span> <span class="c1"># &lt;END&gt; scape num. 2</span>


            <span class="mi">3</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1">#scape name; each scape must be give a unique string or numeric</span>
                <span class="c1">#(e.g. 0, 0.1, &#39;scape0&#39;, &#39;1994&#39;, &#39;mean_T&#39;); default to serial</span>
                <span class="c1">#integers from 0</span>

        <span class="c1">############################</span>
        <span class="c1">#### scape num. 3: init ####</span>
        <span class="c1">############################</span>

                <span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="c1">#initiating parameters for this scape</span>

                    <span class="s1">&#39;file&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#parameters for making a scape from a file (read in</span>
                        <span class="c1">#from a GIS raster file or a numpy txt array file)</span>
                        <span class="s1">&#39;filepath&#39;</span><span class="p">:</span>                     <span class="s1">&#39;/PATH/TO/FILE.EXT&#39;</span><span class="p">,</span>
                            <span class="c1">#filepath to read into this scape</span>
                        <span class="s1">&#39;scale_min_val&#39;</span><span class="p">:</span>                <span class="o">-</span><span class="mf">1.37</span><span class="p">,</span>
                            <span class="c1">#minimum values to use for rescaling the raster (will be</span>
                                <span class="c1">#rescaled to 0&lt;=x&lt;=1); NOTE: this may be different than the actual minimum</span>
                                <span class="c1">#value in the raster, especially if raster will be changing to a future raster</span>
                                <span class="c1">#with values outside the range of this one</span>
                        <span class="s1">&#39;scale_max_val&#39;</span><span class="p">:</span>                <span class="mf">19.11</span>
                            <span class="c1">#maxmimum input value against to which to rescale the raster (will be rescaled to 0&lt;=x&lt;=1)</span>
                        <span class="p">},</span>

                    <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;init&#39;</span>

                <span class="p">},</span> <span class="c1"># &lt;END&gt; scape num. 3</span>



    <span class="c1">#### NOTE: Individual Scapes&#39; sections can be copy-and-pasted (and</span>
    <span class="c1">#### assigned distinct keys and names), to create additional Scapes.</span>


            <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;scapes&#39;</span>

        <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;land&#39;</span>

<span class="c1">###################</span>
<span class="c1">#### COMMUNITY ####</span>
<span class="c1">###################</span>
    <span class="s1">&#39;comm&#39;</span><span class="p">:</span> <span class="p">{</span>

        <span class="s1">&#39;pops&#39;</span><span class="p">:</span> <span class="p">{</span>

            <span class="mi">0</span>  <span class="p">:</span>   <span class="p">{</span>
                <span class="c1">#pop name; each pop must get a unique numeric or string name</span>
                <span class="c1">#(e.g. 0, 0.1, &#39;pop0&#39;, &#39;south&#39;, &#39;C_fasciata&#39;); default to</span>
                <span class="c1">#serial integers from 0</span>

            <span class="c1">##########################</span>
            <span class="c1">#### pop num. 0: init ####</span>
            <span class="c1">##########################</span>

                <span class="s1">&#39;init&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s1">&#39;N&#39;</span><span class="p">:</span>                <span class="mi">200</span><span class="p">,</span>
                        <span class="c1">#starting population size</span>
                    <span class="s1">&#39;K_scape_num&#39;</span><span class="p">:</span>      <span class="mi">0</span><span class="p">,</span>
                        <span class="c1">#the scape_num of the raster to use as the carrying-capacity raster (K)</span>
                    <span class="s1">&#39;K_fact&#39;</span><span class="p">:</span>           <span class="mi">2</span>
                        <span class="c1">#the factor to multiply the K raster by in order to generate pop.K</span>
                    <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;init&#39;</span>

            <span class="c1">############################</span>
            <span class="c1">#### pop num. 0: mating ####</span>
            <span class="c1">############################</span>

                <span class="s1">&#39;mating&#39;</span>    <span class="p">:</span> <span class="p">{</span>
                    <span class="s1">&#39;repro_age&#39;</span><span class="p">:</span>            <span class="mi">0</span><span class="p">,</span>
                        <span class="c1">#age at sexual maturity (int or float for non-sexual species, tuple or list</span>
                            <span class="c1">#of two ints/floats for sexual species; set to &#39;None&#39; to not make this</span>
                            <span class="c1">#an age-structured species</span>
                    <span class="s1">&#39;max_age&#39;</span><span class="p">:</span>              <span class="mi">5</span><span class="p">,</span>
                        <span class="c1">#age beyond which all individuals will automatically die; default to None</span>
                    <span class="s1">&#39;sex&#39;</span><span class="p">:</span>                  <span class="bp">False</span><span class="p">,</span>
                        <span class="c1">#is this a sexual species?</span>
                    <span class="s1">&#39;sex_ratio&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="o">/</span><span class="mi">1</span><span class="p">,</span>
                        <span class="c1">#ratio of males to females</span>
                    <span class="c1">#NOTE: I CAN PROBABLY GET RID OF THIS PARAMETER...</span>
                    <span class="s1">&#39;distweighted_birth&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
                        <span class="c1">#should the probability of birth be weighted by the distance between</span>
                            <span class="c1">#individuals in a pair?</span>
                    <span class="s1">&#39;R&#39;</span><span class="p">:</span>                    <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#pop intrinsic growth rate</span>
                    <span class="s1">&#39;b&#39;</span><span class="p">:</span>                    <span class="mf">0.2</span><span class="p">,</span>
                        <span class="c1">#population intrinsic birth rate (implemented as the probability</span>
                            <span class="c1">#that an identified potential mating pair successfully mates);</span>
                            <span class="c1">#NOTE: this may later need to be re-implemented to allow for spatial</span>
                            <span class="c1">#variation in intrinsic rate (e.g. expression as a raster) and/or for</span>
                            <span class="c1">#density-dependent births as well as deaths</span>
                    <span class="s1">&#39;n_births_distr_lambda&#39;</span><span class="p">:</span>      <span class="mi">4</span><span class="p">,</span>
                        <span class="c1">#expected value of offspring for a successful mating pair (used as the lambda value in a Poisson distribution)</span>
                    <span class="s1">&#39;mating_radius&#39;</span><span class="p">:</span>        <span class="mi">1</span>
                        <span class="c1">#radius of mate-searching area</span>
                    <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;mating&#39;</span>

            <span class="c1">###############################</span>
            <span class="c1">#### pop num. 0: mortality ####</span>
            <span class="c1">###############################</span>

                <span class="s1">&#39;mortality&#39;</span>     <span class="p">:</span> <span class="p">{</span>
                    <span class="s1">&#39;n_deaths_distr_sigma&#39;</span><span class="p">:</span>           <span class="mf">0.2</span><span class="p">,</span>
                        <span class="c1">#std for the normal distribution used to choose the r.v. of deaths</span>
                            <span class="c1">#per timestep (mean of this distribution is the overshoot,</span>
                            <span class="c1">#as calculated from pop.size and pop.census())</span>
                    <span class="s1">&#39;dens_dependent_fitness&#39;</span><span class="p">:</span>   <span class="bp">True</span><span class="p">,</span>
                        <span class="c1">#should fitness be density dependent? (note: helps to avoid subpopulation &#39;clumping&#39;)</span>
                    <span class="s1">&#39;dens_grid_window_width&#39;</span><span class="p">:</span>   <span class="bp">None</span><span class="p">,</span>
                        <span class="c1">#with window-width used for the Density_Grid_Stack that calculates pop density</span>
                            <span class="c1">#(if set to None, defaults to the closest factor of the larger landscape</span>
                            <span class="c1">#dimension to 1/10th of that dimension)</span>
                            <span class="c1">#NOTE: will eventually default to an approximation of Wright&#39;s genetic neighborhood</span>
                            <span class="c1">#distance, based on the population&#39;s movement/dispersal parameters</span>
                    <span class="s1">&#39;d_min&#39;</span><span class="p">:</span>                     <span class="mf">0.01</span><span class="p">,</span>
                        <span class="c1">#minimum neutral (i.e. non-selection driven) probability of death</span>
                    <span class="s1">&#39;d_max&#39;</span><span class="p">:</span>                    <span class="mf">0.90</span><span class="p">,</span>
                        <span class="c1">#maximum neutral probability of death</span>
                    <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;mortality&#39;</span>

            <span class="c1">##############################</span>
            <span class="c1">#### pop num. 0: movement ####</span>
            <span class="c1">##############################</span>

                <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="p">{</span>
                   <span class="s1">&#39;move&#39;</span><span class="p">:</span>          <span class="bp">True</span><span class="p">,</span>
                        <span class="c1">#is this a mobile species?</span>
                    <span class="s1">&#39;direction_distr_mu&#39;</span><span class="p">:</span>     <span class="mi">0</span><span class="p">,</span>
                        <span class="c1">#mu for von mises distribution defining movement directions</span>
                    <span class="s1">&#39;direction_distr_kappa&#39;</span><span class="p">:</span>  <span class="mi">0</span><span class="p">,</span>
                        <span class="c1">#kappa for von mises distribution</span>
                    <span class="s1">&#39;distance_distr_mu&#39;</span><span class="p">:</span>      <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#mean movement-distance (lognormal distribution)</span>
                    <span class="s1">&#39;distance_distr_sigma&#39;</span><span class="p">:</span>   <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#sd of movement distance</span>
                    <span class="s1">&#39;dispersal_distr_mu&#39;</span><span class="p">:</span>     <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#mean dispersal distance (lognormal distribution)</span>
                    <span class="s1">&#39;dispersal_distr_sigma&#39;</span><span class="p">:</span>  <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#sd of dispersal distance</span>

                    <span class="s1">&#39;move_surf&#39;</span>     <span class="p">:</span> <span class="p">{</span>
                        <span class="s1">&#39;scape_num&#39;</span><span class="p">:</span>                    <span class="mi">0</span><span class="p">,</span>
                            <span class="c1">#scape number to use as the movement surface</span>
                        <span class="s1">&#39;mixture&#39;</span><span class="p">:</span>                      <span class="bp">True</span><span class="p">,</span>
                            <span class="c1">#should this MovementSurface be composed of</span>
                            <span class="c1">#VonMises mixture distribution approximations (i.e.</span>
                            <span class="c1">#True) or of unimodal VonMises distribution</span>
                            <span class="c1">#approximations (i.e. False); the latter is better</span>
                            <span class="c1">#for landscapes characterized by gradual gradients</span>
                            <span class="c1">#(i.e. where many cells&#39; neighborhoods have very</span>
                            <span class="c1">#little variation between the highest- and lowest</span>
                            <span class="c1">#permeability values, so that mixture distributions</span>
                            <span class="c1">#would be largely uniform in all directions);</span>
                            <span class="c1">#defaults to True</span>
                        <span class="s1">&#39;approximation_len&#39;</span><span class="p">:</span>            <span class="mi">7500</span><span class="p">,</span>
                            <span class="c1">#length of the lookup vectors (numpy arrays) used to approximate</span>
                                <span class="c1">#the VonMises mixture distributions at each cell</span>
                        <span class="s1">&#39;vm_distr_kappa&#39;</span><span class="p">:</span>                     <span class="bp">None</span><span class="p">,</span>
                            <span class="c1">#kappa value to use in the von Mises mixture distributions (KDEs)</span>
                                <span class="c1">#underlying resistance surface movement</span>
                        <span class="s1">&#39;gauss_KDE_bw&#39;</span><span class="p">:</span>                 <span class="bp">None</span>
                            <span class="c1">#bandwidth value to use in the Gaussian KDEs that are created to</span>
                                <span class="c1">#approximate the von Mises mixture distributions (KDEs)</span>
                                <span class="c1">#underlying resistance surface movement</span>
                        <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;move_surf&#39;</span>

                    <span class="p">},</span>    <span class="c1"># &lt;END&gt; &#39;movement&#39;</span>


            <span class="c1">##############################</span>
            <span class="c1">#### pop num. 0: gen_arch ####</span>
            <span class="c1">##############################</span>

                <span class="s1">&#39;gen_arch&#39;</span><span class="p">:</span> <span class="p">{</span>
                    <span class="s1">&#39;L&#39;</span><span class="p">:</span>                        <span class="mi">10</span><span class="p">,</span>
                        <span class="c1">#total number of loci</span>
                    <span class="s1">&#39;l_c&#39;</span><span class="p">:</span>                      <span class="p">[</span><span class="mi">10</span><span class="p">],</span>
                        <span class="c1">#chromosome lengths [sum(l_c) == L is enforced]</span>
                    <span class="s1">&#39;gen_arch_file&#39;</span><span class="p">:</span>            <span class="bp">None</span><span class="p">,</span>
                        <span class="c1">#if not None, should point to a file stipulating a</span>
                            <span class="c1">#custom genomic architecture (i.e. a CSV with loci</span>
                            <span class="c1">#as rows and &#39;locus_num&#39;, &#39;p&#39;, &#39;dom&#39;, &#39;r&#39;, &#39;trait&#39;,</span>
                            <span class="c1">#and &#39;alpha&#39; as columns, such as is created by</span>
                            <span class="c1">#main.make_parameters_file, when the custom_gen_arch</span>
                            <span class="c1">#arugment is True)</span>
                    <span class="s1">&#39;mu_neut&#39;</span><span class="p">:</span>                  <span class="mf">1e-9</span><span class="p">,</span>
                        <span class="c1">#genome-wide neutral mutation rate, per base per generation</span>
                            <span class="c1">#(set to 0 to disable neutral mutation)</span>
                    <span class="s1">&#39;mu_delet&#39;</span><span class="p">:</span>                 <span class="mi">0</span><span class="p">,</span>
                        <span class="c1">#genome-wide deleterious mutation rate, per base per generation</span>
                            <span class="c1">#(set to 0 to disable deleterious mutation)</span>
                            <span class="c1">#NOTE: these mutations will fall outside the loci involved in any traits</span>
                            <span class="c1">#being simulated, and are simply treated as universally deleterious, with the same</span>
                            <span class="c1">#negative influence on fitness regardless of spatial context</span>
                    <span class="s1">&#39;mut_log&#39;</span><span class="p">:</span>                  <span class="bp">False</span><span class="p">,</span>
                        <span class="c1">#whether or not to store a mutation log; if true, will be saved as mut_log.txt</span>
                        <span class="c1">#within each iteration&#39;s subdirectory</span>
                    <span class="s1">&#39;delet_s_distr_shape&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
                    <span class="s1">&#39;delet_s_distr_scale&#39;</span><span class="p">:</span>      <span class="mf">0.2</span><span class="p">,</span>
                        <span class="c1">#mean and standard deviation of the gamma distribution</span>
                        <span class="c1">#parameterizig the per-allele effect size of</span>
                        <span class="c1">#deleterious mutations (std = 0 will fix all mutations</span>
                        <span class="c1">#for the mean value)</span>
                    <span class="s1">&#39;r_distr_alpha&#39;</span><span class="p">:</span>            <span class="mf">0.5</span><span class="p">,</span>
                        <span class="c1">#alpha for beta distribution of linkage values</span>
                            <span class="c1">#NOTE: alpha = 14.999e9, beta = 15e9 has a VERY sharp peak on D = 0.4998333,</span>
                            <span class="c1">#with no values exceeding equalling or exceeding 0.5 in 10e6 draws in R</span>
                    <span class="s1">&#39;r_distr_beta&#39;</span><span class="p">:</span>             <span class="mf">15e9</span><span class="p">,</span>
                        <span class="c1">#beta for beta distribution of linkage values</span>
                    <span class="s1">&#39;dom&#39;</span><span class="p">:</span>                      <span class="bp">False</span><span class="p">,</span>
                        <span class="c1">#whether or not loci should be dominant</span>
                        <span class="c1">#(if True, the 1 allele will be dominant at each locus;</span>
                        <span class="c1">#if False, all loci will be codominant; defaults to False)</span>
                    <span class="s1">&#39;pleiotropy&#39;</span><span class="p">:</span>               <span class="bp">True</span><span class="p">,</span>
                        <span class="c1">#allow pleiotropy? (i.e. allow same locus to affect value of more than one trait?) false</span>
                    <span class="s1">&#39;recomb_rate_custom_fn&#39;</span><span class="p">:</span>    <span class="bp">None</span><span class="p">,</span>
                        <span class="c1">#if provided, must be a function that returns a single recombination rate value (r) when called</span>
                    <span class="s1">&#39;recomb_lookup_array_size&#39;</span><span class="p">:</span> <span class="nb">int</span><span class="p">(</span><span class="mf">1e3</span><span class="p">),</span>
                        <span class="c1">#the size of the recombination-path lookup array to have</span>
                            <span class="c1">#read in at one time (needs to be comfortably larger than the anticipated totaly number of</span>
                            <span class="c1">#recombination paths to be drawn at once, i.e. than 2 times the anticipated most number of births at once)</span>
                    <span class="s1">&#39;n_recomb_paths&#39;</span><span class="p">:</span>           <span class="nb">int</span><span class="p">(</span><span class="mf">1e4</span><span class="p">),</span>
                        <span class="c1">#the total number of distinct recombination paths to</span>
                            <span class="c1">#generate at the outset, to approximate truly free recombination at the recombination rates specified</span>
                            <span class="c1">#by the genomic architecture (hence the larger the value the less the likelihood of mis-approximation artifacts)</span>

                    <span class="s1">&#39;traits&#39;</span><span class="p">:</span> <span class="p">{</span>
      <span class="c1">#########</span>
                        <span class="c1">#trait 0#</span>
                        <span class="c1">#########</span>
                        <span class="mi">0</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#trait name; each trait must get a unique numeric or</span>
                        <span class="c1">#string name (e.g. 0, &#39;trt0&#39;, &#39;bill_length&#39;); trait</span>
                        <span class="c1">#names default to serial integers from 0</span>
                            <span class="s1">&#39;scape_num&#39;</span><span class="p">:</span>        <span class="mi">2</span><span class="p">,</span>
                                <span class="c1">#the landscape numbers to be used for selection on this trait</span>
                            <span class="s1">&#39;phi&#39;</span><span class="p">:</span>              <span class="mf">0.1</span><span class="p">,</span>
                                <span class="c1">#phenotypic selection coefficient for this trait; can either be a</span>
                                    <span class="c1">#numeric value, or can be an array of spatialized selection</span>
                                    <span class="c1">#values (with dimensions equal to land.dims)</span>
                            <span class="s1">&#39;n_loci&#39;</span><span class="p">:</span>           <span class="mi">1</span><span class="p">,</span>
                                <span class="c1">#number of loci to be assigned to this trait</span>
                            <span class="s1">&#39;mu&#39;</span><span class="p">:</span>      <span class="mf">1e-9</span><span class="p">,</span>
                                <span class="c1">#mutation rate for this trait (if set to 0, or if genome[&#39;mutation&#39;] == False, no mutation will occur)</span>
                                    <span class="c1">#(set to 0 to disable mutation for this trait)</span>
                            <span class="s1">&#39;alpha_distr_mu&#39;</span> <span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                            <span class="s1">&#39;alpha_distr_sigma&#39;</span> <span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
                                <span class="c1">#the mean and standard deviation of the normal distribution used to choose effect size</span>
                                    <span class="c1">#(alpha) for this trait&#39;s loci</span>
                                    <span class="c1">#NOTE: for mean = 0, std = 0.5, one average locus is enough to generate both optimum</span>
                                    <span class="c1">#genotypes; for mean = 0, std = 0.025, 10 loci should generate both (on average, but depends of course on</span>
                                    <span class="c1">#the random sample of alphas drawn); and so on linearly</span>
                            <span class="s1">&#39;gamma&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="p">,</span>
                                <span class="c1">#gamma exponent for the trait&#39;s fitness function (determines the shape of the</span>
                                <span class="c1">#curve of fitness as a function of absolute difference between an individual&#39;s</span>
                                <span class="c1">#phenotype and its environment; &lt;1 = concave up, 1 = linear, &gt;1 = convex up)</span>
                            <span class="s1">&#39;univ_adv&#39;</span><span class="p">:</span>      <span class="bp">False</span>
                                <span class="c1">#is the trait universally advantageous? if so, phenotypes closer to 1 will</span>
                                    <span class="c1">#have higher fitness at all locations on the land</span>
                            <span class="p">},</span> <span class="c1"># &lt;END&gt; trait 0</span>
      <span class="c1">#########</span>
                        <span class="c1">#trait 1#</span>
                        <span class="c1">#########</span>
                        <span class="mi">1</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#trait name; each trait must get a unique numeric or</span>
                        <span class="c1">#string name (e.g. 0, &#39;trt0&#39;, &#39;bill_length&#39;); trait</span>
                        <span class="c1">#names default to serial integers from 0</span>
                            <span class="s1">&#39;scape_num&#39;</span><span class="p">:</span>        <span class="mi">2</span><span class="p">,</span>
                                <span class="c1">#the landscape numbers to be used for selection on this trait</span>
                            <span class="s1">&#39;phi&#39;</span><span class="p">:</span>              <span class="mf">0.1</span><span class="p">,</span>
                                <span class="c1">#phenotypic selection coefficient for this trait; can either be a</span>
                                    <span class="c1">#numeric value, or can be an array of spatialized selection</span>
                                    <span class="c1">#values (with dimensions equal to land.dims)</span>
                            <span class="s1">&#39;n_loci&#39;</span><span class="p">:</span>           <span class="mi">1</span><span class="p">,</span>
                                <span class="c1">#number of loci to be assigned to this trait</span>
                            <span class="s1">&#39;mu&#39;</span><span class="p">:</span>      <span class="mf">1e-9</span><span class="p">,</span>
                                <span class="c1">#mutation rate for this trait (if set to 0, or if genome[&#39;mutation&#39;] == False, no mutation will occur)</span>
                                    <span class="c1">#(set to 0 to disable mutation for this trait)</span>
                            <span class="s1">&#39;alpha_distr_mu&#39;</span> <span class="p">:</span> <span class="mi">0</span><span class="p">,</span>
                            <span class="s1">&#39;alpha_distr_sigma&#39;</span> <span class="p">:</span> <span class="mf">0.5</span><span class="p">,</span>
                                <span class="c1">#the mean and standard deviation of the normal distribution used to choose effect size</span>
                                    <span class="c1">#(alpha) for this trait&#39;s loci</span>
                                    <span class="c1">#NOTE: for mean = 0, std = 0.5, one average locus is enough to generate both optimum</span>
                                    <span class="c1">#genotypes; for mean = 0, std = 0.025, 10 loci should generate both (on average, but depends of course on</span>
                                    <span class="c1">#the random sample of alphas drawn); and so on linearly</span>
                            <span class="s1">&#39;gamma&#39;</span><span class="p">:</span>            <span class="mi">1</span><span class="p">,</span>
                                <span class="c1">#gamma exponent for the trait&#39;s fitness function (determines the shape of the</span>
                                <span class="c1">#curve of fitness as a function of absolute difference between an individual&#39;s</span>
                                <span class="c1">#phenotype and its environment; &lt;1 = concave up, 1 = linear, &gt;1 = convex up)</span>
                            <span class="s1">&#39;univ_adv&#39;</span><span class="p">:</span>      <span class="bp">False</span>
                                <span class="c1">#is the trait universally advantageous? if so, phenotypes closer to 1 will</span>
                                    <span class="c1">#have higher fitness at all locations on the land</span>
                            <span class="p">},</span> <span class="c1"># &lt;END&gt; trait 1</span>


    <span class="c1">#### NOTE: Individual Traits&#39; sections can be copy-and-pasted (and</span>
    <span class="c1">#### assigned distinct keys and names), to create additional Traits.</span>


                        <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;traits&#39;</span>

                    <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;gen_arch&#39;</span>


            <span class="c1">############################</span>
            <span class="c1">#### pop num. 0: change ####</span>
            <span class="c1">############################</span>

                <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="p">{</span>

                    <span class="s1">&#39;dem&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#(all population sizes are expressed relative to the carrying-capacity</span>
                            <span class="c1">#raster at the time that the demographic change event begins (i.e. as</span>
                            <span class="c1">#factors by which pop.K will be multiplied; thus they can also be thought</span>
                            <span class="c1">#of as multipliers of the expected total population size (i.e. pop.K.sum())</span>
                            <span class="c1">#and they will generally change the average population size by that multiple,</span>
                            <span class="c1">#but of course not precisely, because population size is stochastic. If you</span>
                            <span class="c1">#seek exact control of total population size, please seek a simpler simulation</span>
                            <span class="c1">#model, perhaps a coalescent one.</span>

                        <span class="mi">0</span><span class="p">:</span> <span class="p">{</span>
                            <span class="c1">#can add an arbitrary number of demographic change events for</span>
                                <span class="c1">#each population, each event identified by a distinct integer</span>
                            <span class="s1">&#39;kind&#39;</span><span class="p">:</span>             <span class="s1">&#39;custom&#39;</span><span class="p">,</span>
                                <span class="c1">#what kind of change event? (&#39;monotonic&#39;, &#39;stochastic&#39;, &#39;cyclical&#39;, &#39;custom&#39;)</span>
                            <span class="s1">&#39;start&#39;</span><span class="p">:</span>            <span class="mi">200</span><span class="p">,</span>
                                <span class="c1">#at which timestep should the event start?</span>
                            <span class="s1">&#39;end&#39;</span><span class="p">:</span>              <span class="mi">1200</span><span class="p">,</span>
                                <span class="c1">#at which timestep should the event end?</span>
                            <span class="s1">&#39;rate&#39;</span><span class="p">:</span>             <span class="o">.</span><span class="mi">98</span><span class="p">,</span>
                                <span class="c1">#at what rate should the population change each timestep</span>
                                    <span class="c1">#(positive for growth, negative for reduction)</span>
                            <span class="s1">&#39;interval&#39;</span><span class="p">:</span>         <span class="mi">11</span><span class="p">,</span>
                                <span class="c1">#at what interval should stochastic change take place (None defaults to every timestep)</span>
                            <span class="s1">&#39;dist&#39;</span><span class="p">:</span>             <span class="s1">&#39;uniform&#39;</span><span class="p">,</span>
                                <span class="c1">#what distribution to draw stochastic population sizes from (valid values: &#39;uniform&#39;, &#39;normal&#39;)</span>
                            <span class="s1">&#39;size_target&#39;</span><span class="p">:</span>      <span class="bp">None</span><span class="p">,</span>
                                <span class="c1">#what is the target size of the demographic change event (defaults to None)</span>
                            <span class="s1">&#39;n_cycles&#39;</span><span class="p">:</span>         <span class="mi">20</span><span class="p">,</span>
                                <span class="c1">#how many cycles of cyclical change should occur during the event?</span>
                            <span class="s1">&#39;size_range&#39;</span><span class="p">:</span>       <span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="mf">1.5</span><span class="p">),</span>
                                <span class="c1">#an iterable of the min and max population sizes to be used in stochastic or cyclical changes</span>
                            <span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="p">[</span><span class="mi">6</span><span class="p">,</span><span class="mi">8</span><span class="p">],</span>
                                <span class="c1">#at which timesteps should custom changes take place?</span>
                            <span class="s1">&#39;sizes&#39;</span><span class="p">:</span>            <span class="p">[</span><span class="mi">2</span><span class="p">,</span><span class="mf">0.25</span><span class="p">],</span>
                                <span class="c1">#what custom size-changes should occur at the above-stipulated timesteps?</span>
                            <span class="p">}</span> <span class="c1"># &lt;END&gt; event 0</span>




    <span class="c1">#### NOTE: Individual demographic change events&#39; sections can be</span>
    <span class="c1">#### copy-and-pasted (and assigned distinct keys and names), to create</span>
    <span class="c1">#### additional events.</span>


                        <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;dem&#39;</span>

                    <span class="s1">&#39;parameters&#39;</span><span class="p">:</span> <span class="p">{</span>
                        <span class="c1">#other (i.e. non-demographic) population change events</span>
                        <span class="s1">&#39;b&#39;</span><span class="p">:</span> <span class="p">{</span>
                            <span class="c1">#the life-history parameters to be changed should be the keys in this dict,</span>
                                <span class="c1">#and values are dictionaries containing a list of timesteps</span>
                                <span class="c1">#at which to changed those parameters and a list of values</span>
                                <span class="c1">#to which to change them</span>
                            <span class="s1">&#39;timesteps&#39;</span><span class="p">:</span>        <span class="bp">None</span><span class="p">,</span>
                            <span class="s1">&#39;vals&#39;</span><span class="p">:</span>           <span class="bp">None</span>
                                <span class="p">}</span>


    <span class="c1">#### NOTE: Individual life-history paramter change events&#39; sections can be</span>
    <span class="c1">#### copy-and-pasted (and assigned distinct keys and names), to create</span>
    <span class="c1">#### additional events.</span>


                            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;parameters&#39;</span>

                        <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;change&#39;</span>

                <span class="p">},</span> <span class="c1"># &lt;END&gt; pop num. 0</span>



    <span class="c1">#### NOTE: Individual Populations&#39; sections can be copy-and-pasted (and</span>
    <span class="c1">#### assigned distinct keys and names), to create additional Populations.</span>


            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;pops&#39;</span>

        <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;comm&#39;</span>

<span class="c1">###############</span>
<span class="c1">#### MODEL ####</span>
<span class="c1">###############</span>
    <span class="s1">&#39;model&#39;</span><span class="p">:</span> <span class="p">{</span>
        <span class="s1">&#39;time&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1">#parameters to control the number of burn-in and main timesteps to</span>
            <span class="c1">#run for each iterations</span>
            <span class="s1">&#39;T&#39;</span><span class="p">:</span>            <span class="mi">100</span><span class="p">,</span>
                <span class="c1">#total model runtime (in timesteps)</span>
            <span class="s1">&#39;burn_T&#39;</span><span class="p">:</span>       <span class="mi">30</span>
                <span class="c1">#minimum burn-in runtime (in timesteps; this is a mininimum because</span>
                    <span class="c1">#burn-in will run for at least this long but until</span>
                    <span class="c1">#stationarity detected, which will likely be longer)</span>
            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;timesteps&#39;</span>

        <span class="s1">&#39;its&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1">#parameters to control how many iterations of the model to run,</span>
            <span class="c1">#and whether or not to randomize the land and/or community</span>
            <span class="c1">#objects in each model iteration</span>
            <span class="s1">&#39;n_its&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                <span class="c1">#how many iterations of the model should be run?</span>
            <span class="s1">&#39;rand_land&#39;</span><span class="p">:</span>    <span class="bp">False</span><span class="p">,</span>
                <span class="c1">#randomize the land for each new iteration?</span>
            <span class="s1">&#39;rand_comm&#39;</span><span class="p">:</span>    <span class="bp">False</span><span class="p">,</span>
                <span class="c1">#randomize the community for each new iteration?</span>
            <span class="s1">&#39;rand_burn&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
                <span class="c1">#randomize the burn-in for each new iteration? (i.e. burn in</span>
                <span class="c1">#each time, or burn in once at creation and then use the same</span>
                <span class="c1">#burnt-in population for each iteration?)</span>
            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;iterations&#39;</span>


        <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1">#dictionary defining the data to be collected, the sampling</span>
            <span class="c1">#strategy to use, the timesteps for collection, and other parameters</span>
            <span class="s1">&#39;sampling&#39;</span><span class="p">:</span> <span class="p">{</span>
                <span class="c1">#args to be unpacked into sampling function (see docstring</span>
                    <span class="c1">#of sample_data function in data module for details)</span>
                <span class="s1">&#39;scheme&#39;</span><span class="p">:</span>               <span class="s1">&#39;random&#39;</span><span class="p">,</span>
                    <span class="c1">#valid: &#39;all&#39;, &#39;random&#39;, &#39;point&#39;, or &#39;transect&#39;</span>
                <span class="s1">&#39;n&#39;</span><span class="p">:</span>                    <span class="mi">50</span><span class="p">,</span>
                    <span class="c1">#size of samples to be collected (in number of individuals)</span>
                <span class="s1">&#39;points&#39;</span><span class="p">:</span>               <span class="bp">None</span><span class="p">,</span>
                    <span class="c1">#the x,y points at which data should be sampled (expressed</span>
                        <span class="c1">#as a list or tuple of length-2 lists or 2-tuples)</span>
                <span class="s1">&#39;transect_endpoints&#39;</span><span class="p">:</span>   <span class="bp">None</span><span class="p">,</span>
                    <span class="c1">#endpoints of the transect to be sampled (only needed if</span>
                        <span class="c1">#scheme is &#39;transect), expressed as a pair of</span>
                        <span class="c1">#ordered x,y pairs (in tuples or lists)</span>
                <span class="s1">&#39;n_transect_points&#39;</span><span class="p">:</span>    <span class="bp">None</span><span class="p">,</span>
                    <span class="c1">#the number of evenly spaced points along the transect</span>
                    <span class="c1">#at which to sample (only needed if scheme is &#39;transect&#39;)</span>
                <span class="s1">&#39;radius&#39;</span><span class="p">:</span>               <span class="bp">None</span><span class="p">,</span>
                    <span class="c1">#radius around sampling points within which to sample</span>
                    <span class="c1">#individuals (only needed is scheme is &#39;point&#39; or</span>
                    <span class="c1">#&#39;transect&#39;)</span>
                <span class="s1">&#39;when&#39;</span><span class="p">:</span>                 <span class="bp">None</span><span class="p">,</span>
                    <span class="c1">#can be an integer (in which case data will be collected every</span>
                    <span class="c1">#that many timesteps, plus at the end) or a list of specific</span>
                    <span class="c1">#timesteps; a value of 0 or None will default to a single</span>
                    <span class="c1">#data-collection step after the model has run</span>
                <span class="s1">&#39;include_land&#39;</span><span class="p">:</span>         <span class="bp">False</span><span class="p">,</span>
                    <span class="c1">#if True, will save the Land object each time other data is saved</span>
                    <span class="c1">#(probably only useful if land is changing in some way not manually coded by the user)</span>
                <span class="s1">&#39;include_fixed_sites&#39;</span><span class="p">:</span>  <span class="bp">False</span><span class="p">,</span>
                    <span class="c1">#if True, and if genetic data is to be formatted as VCFs,</span>
                        <span class="c1">#the VCFs will contain fixed sites, not just variants</span>
                        <span class="c1">#(defaults to False)</span>
                <span class="p">},</span>
            <span class="s1">&#39;format&#39;</span><span class="p">:</span> <span class="p">{</span>
                <span class="s1">&#39;gen_format&#39;</span><span class="p">:</span>           <span class="p">[</span><span class="s1">&#39;vcf&#39;</span><span class="p">,</span> <span class="s1">&#39;fasta&#39;</span><span class="p">],</span>
                    <span class="c1">#format to use for saving genetic data;</span>
                        <span class="c1">#currently valid values: &#39;vcf&#39;, &#39;fasta&#39;,</span>
                        <span class="c1">#or a list containing both, if both</span>
                        <span class="c1">#should be written</span>
                <span class="s1">&#39;geo_vect_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;csv&#39;</span><span class="p">,</span>
                    <span class="c1">#format to use for saving geographic points;</span>
                        <span class="c1">#currently valid values: &#39;csv&#39;, &#39;shapefile&#39;, &#39;geojson&#39;</span>
                <span class="s1">&#39;geo_rast_format&#39;</span><span class="p">:</span>      <span class="s1">&#39;geotiff&#39;</span><span class="p">,</span>
                    <span class="c1">#format to use for saving landscape rasters (which will</span>
                        <span class="c1">#only be saved if the &#39;include_land&#39; parameter in the</span>
                        <span class="c1">#sampling subdict is True);</span>
                        <span class="c1">#currently valid values: &#39;geotiff&#39;, &#39;txt&#39;</span>
                <span class="p">},</span>
            <span class="p">},</span> <span class="c1">#&lt;END&gt; &#39;data&#39;</span>


        <span class="s1">&#39;stats&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1">#dictionary defining which stats to be calculated, and parameters for</span>
                <span class="c1">#their calculation (including frequency, in timesteps, of collection)</span>
                <span class="c1">#valid stats include:</span>
                    <span class="c1"># &#39;Nt&#39;  : population census size</span>
                    <span class="c1"># &#39;het&#39; : heterozygosity</span>
                    <span class="c1"># &#39;maf&#39; : minor allele frequency</span>
                    <span class="c1"># &#39;ld&#39;  : linkage disequilibrium</span>
                    <span class="c1"># &#39;mean_fit&#39; : mean fitness</span>
            <span class="s1">&#39;Nt&#39;</span><span class="p">:</span>       <span class="p">{</span><span class="s1">&#39;calc&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                         <span class="s1">&#39;freq&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
                        <span class="p">},</span>
            <span class="s1">&#39;het&#39;</span><span class="p">:</span>      <span class="p">{</span><span class="s1">&#39;calc&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                         <span class="s1">&#39;freq&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">,</span>
                         <span class="s1">&#39;mean&#39;</span><span class="p">:</span> <span class="bp">False</span><span class="p">,</span>
                        <span class="p">},</span>
            <span class="s1">&#39;maf&#39;</span><span class="p">:</span>      <span class="p">{</span><span class="s1">&#39;calc&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                         <span class="s1">&#39;freq&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
                        <span class="p">},</span>
            <span class="s1">&#39;ld&#39;</span><span class="p">:</span>       <span class="p">{</span><span class="s1">&#39;calc&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                         <span class="s1">&#39;freq&#39;</span><span class="p">:</span> <span class="mi">10</span><span class="p">,</span>
                        <span class="p">},</span>
            <span class="s1">&#39;mean_fit&#39;</span><span class="p">:</span> <span class="p">{</span><span class="s1">&#39;calc&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">,</span>
                         <span class="s1">&#39;freq&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">,</span>
                        <span class="p">},</span>
            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;stats&#39;</span>


        <span class="s1">&#39;seed&#39;</span><span class="p">:</span> <span class="p">{</span>
            <span class="c1">#parameters to control whether and how to set the seed</span>
            <span class="s1">&#39;set&#39;</span><span class="p">:</span>          <span class="bp">True</span><span class="p">,</span>
                <span class="c1">#set the seed? (for reproducibility)</span>
            <span class="s1">&#39;num&#39;</span><span class="p">:</span>          <span class="mi">94618</span>
                <span class="c1">#value used to seed random number generators</span>
            <span class="p">},</span> <span class="c1"># &lt;END&gt; &#39;seed&#39;</span>

        <span class="p">}</span> <span class="c1"># &lt;END&gt; &#39;model&#39;</span>

    <span class="p">}</span> <span class="c1"># &lt;END&gt; params</span>
</pre></div>
</div>
</div>
</div>
<hr class="docutils" />
<div class="section" id="class-and-function-docstrings">
<h3>Class and function docstrings<a class="headerlink" href="#class-and-function-docstrings" title="Permalink to this headline">¶</a></h3>
<div class="section" id="gnx-make-parameters-file">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">gnx</span></span><span class="operator"><span class="pre">.</span></span><span class="name"><span class="pre">make_parameters_file</span></span></code><a class="headerlink" href="#gnx-make-parameters-file" title="Permalink to this headline">¶</a></h4>
<p>Create a new parameters file.</p>
<p>Write to disk a new, template parameters file. The file will contain the
numbers and types of sections indicated by the parameters fed to this
function. It can often be used ‘out of the box’ to make a new Model
object, but typically it will be edited by the user to stipulate
the scenario being simulated, then used to instantiate a Model.</p>
<div class="section" id="id1">
<h5>Parameters<a class="headerlink" href="#id1" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>filepath <span class="classifier-delimiter">:</span> <span class="classifier">str, optional</span></dt>
<dd>Where to write the resulting parameters file, in /path/to/filename.py
format. Defaults to None. If None, a file named
“GEONOMICS_params_&lt;datetime&gt;.py” will be written to the working
directory.</dd>
<dt>scapes <span class="classifier-delimiter">:</span> <span class="classifier">{int, list of dicts}, optional</span></dt>
<dd><p class="first">Number (and optionally, types) of Scape-parameter sections to include
in the parameters file that is generated. Defaults to 1. Valid values
and their associated behaviors are:</p>
<dl class="last docutils">
<dt>int:</dt>
<dd><p class="first">Add sections for the stipulated number of Scapes, each with default
settings:</p>
<blockquote class="last">
<div><ul class="simple">
<li>parameters for creating Scapes of type ‘random’ (i.e.
Scapes that will be generated by interpolation from
randomly valued random points)</li>
<li>no ScapeChanger parameters</li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt>
<dd>Each dict in this list should be of the form:
{‘type’:    ‘random’, ‘defined’, ‘file’, or ‘nlmpy’,
‘change’:   bool
}
This will add one section of Scape parameters, with the
contents indicated, for each dict in this list.</dd>
</dl>
</dd>
<dt>populations <span class="classifier-delimiter">:</span> <span class="classifier">{int, list of dicts}, optional</span></dt>
<dd><p class="first">Number (and optionally, types) of Population-parameter sections to
include in the parameters file that is generated. Defaults to 1. Valid
values and their associated behaviors are:</p>
<dl class="last docutils">
<dt>int:</dt>
<dd><p class="first">Add sections for the stipulated number of Populations, each with
default settings:</p>
<blockquote class="last">
<div><ul class="simple">
<li>parameters for movement without a MovementSurface</li>
<li>parameters for a GenomicArchitecture with 0 Traits (i.e. with
only neutral loci)</li>
<li>no PopulationChanger parameters</li>
</ul>
</div></blockquote>
</dd>
<dt>[dict, …, dict]:</dt>
<dd>Each dict should contain at least one argument from among the
following:
{‘movement’:                       bool,
‘movement_surface’:                bool,
‘genomes’:                         bool,
‘n_traits’:                        int,
‘custom_genomic_architecture’:     bool,
‘demographic_change’:              int,
‘parameter_change’:                bool
}
This will add one section of Population parameters, customized
as indicated, for each dict in the list.</dd>
</dl>
</dd>
<dt>data <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first">Whether to include a Data-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="last docutils">
<dt>None, False:</dt>
<dd>Will not add a section for parameterizing data to be collected.
No DataCollector will be created for the Model object made from
the resulting parameters file, and no data will be collected
during the model runs.</dd>
<dt>True:</dt>
<dd>Will add a section that can be used to parameterize which
data will be collected during the model runs, when, and what
file formats will be used to write it to disk.
(This which will be managed by the model’s DataCollector
object.)</dd>
</dl>
</dd>
<dt>stats <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first">Whether to include a Stats-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="last docutils">
<dt>None, False:</dt>
<dd>Will not add a section for parameterizing the statistics to be
calculated. No StatsCollector will be created for the Model
object made from the resulting parameters file, and no
statistics will be calculated during the model runs.</dd>
<dt>True:</dt>
<dd>Will add a section that can be used to parameterize which
statistics will be calculated during the model runs, and when.
(This will be managed by the model’s StatsCollector object.)</dd>
</dl>
</dd>
<dt>seed <span class="classifier-delimiter">:</span> <span class="classifier">bool, optional</span></dt>
<dd><p class="first">Whether to include a seed-parameter section in the parameters file that
is generated. Defaults to None. Valid values and their associated
behaviors are:</p>
<dl class="last docutils">
<dt>None, False:</dt>
<dd>Will not add a section for parameterizing how the random number
generators are seeded, hence generators will be seeded at
random and results will be unreproducible.</dd>
<dt>True:</dt>
<dd>Will add a section for parameterizing how the random number
generators are seeded, so that results will be reproducible.</dd>
</dl>
</dd>
</dl>
<div class="section" id="returns">
<h6>Returns<a class="headerlink" href="#returns" title="Permalink to this headline">¶</a></h6>
<dl class="docutils">
<dt>out <span class="classifier-delimiter">:</span> <span class="classifier">None</span></dt>
<dd>Returns no output. Resulting parameters file will be written to the
location and filename indicated (or by default, will be written to a
file named “GEONOMICS_params_&lt;datetime&gt;.py” in the working directory).</dd>
</dl>
</div>
<div class="section" id="see-also">
<h6>See Also<a class="headerlink" href="#see-also" title="Permalink to this headline">¶</a></h6>
<p>sim.params.make_parameters_file</p>
</div>
<div class="section" id="notes">
<h6>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h6>
<p>All parameters of this function are optional. Calling the function without
providing any parameters will always produce the parameters file for the
default model scenario. This file can be instantiated as a Model object and
run without being edited. Those three steps (create default parameters file;
create model from that parameters file; run the model) serve as a base case
to test successful package installation, and are wrapped around by the
convenience function <cite>gnx.run_default_model</cite>.</p>
</div>
<div class="section" id="examples">
<h6>Examples<a class="headerlink" href="#examples" title="Permalink to this headline">¶</a></h6>
<p>In the simplest example, we can create a parameters file for the default
model. Then (assuming it is the only Geonomics parameters file in the
current working directory, so that it can be unambiguously identified) we
can call the gnx.make_model function to create a Model object from that
file, and then call the Model.run method to run the model (setting the
‘verbose’ parameter to True, so that we can observe model output).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span> <span class="o">=</span> <span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">mod</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">verbose</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
<span class="go">TODO: PUT TYPICAL MODEL OUTPUT HERE, EVEN THOUGH IT&#39;S ONLY PRINTED?</span>
</pre></div>
</div>
<p>We can use some of the function’s arguments, to create a parameters
file for a model with 3 Scapes and 1 Population (all with the default
components for their sections of the parameters file) and with a section
for parameterizing data collection.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span><span class="n">scapes</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span> <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
<p>As a more complex example that is likely to be similar to most use cases,
we can create a parameters file for a model scenario with:</p>
<blockquote>
<div><ul class="simple">
<li>2 Scapes (one being an nlmpy Scape that will not change over model
time, the other being a raster read in from a GIS file and being
subject to change over model time);</li>
<li>2 Populations (the first having genomes, 2 Traits, and movement
that is dictated by a MovementSurface; the second not having
genomes but having a MovementSurface as well, and undergoing
demographic change)</li>
<li>data-collection;</li>
<li>stats-collection;</li>
<li>a section for setting the seed for the random-number generators.</li>
</ul>
</div></blockquote>
<p>We can save this to a file named “2-pop_2-trait_model.py” in our current
working directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_parameters_file</span><span class="p">(</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Scape&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">scapes</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;nlmpy&#39;</span><span class="p">},</span>                              <span class="c1">#scape 1</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;type&#39;</span><span class="p">:</span> <span class="s1">&#39;gis&#39;</span><span class="p">,</span>                                 <span class="c1">#scape 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#list of 2 dicts, each containing the values for each Population&#39;s</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#parameters section</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">populations</span> <span class="o">=</span> <span class="p">[</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>                               <span class="c1">#pop 1</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;n_traits&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">},</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">{</span><span class="s1">&#39;genomes&#39;</span><span class="p">:</span> <span class="kc">False</span><span class="p">,</span>                              <span class="c1">#pop 2</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;movement_surface&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>         <span class="s1">&#39;demographic_change&#39;</span><span class="p">:</span> <span class="kc">True</span><span class="p">}</span>
<span class="gp">&gt;&gt;&gt; </span>        <span class="p">],</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#arguments to the data, stats,and seed parameters</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">data</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stats</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">seed</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="c1">#destination to which to write the resulting parameter file</span>
<span class="gp">&gt;&gt;&gt; </span>    <span class="n">filepath</span> <span class="o">=</span> <span class="s1">&#39;2-pop_2-trait_model.py&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="read-parameters-file">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">read_parameters_file</span></span></code><a class="headerlink" href="#read-parameters-file" title="Permalink to this headline">¶</a></h4>
<p>Create a new ParametersDict object.</p>
<p>Read the Geonomics parameters file saved at the location indicated by
‘filepath’, check its validity (i.e. that all the Scapes and Populations
parameterized in that file have been given distinct names), then use the
file to instantiate a ParametersDict object.</p>
<div class="section" id="id2">
<h5>Parameters<a class="headerlink" href="#id2" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>filepath <span class="classifier-delimiter">:</span> <span class="classifier">str</span></dt>
<dd>String indicating the location of the Geonomics parameters file that
should be made into a ParametersDict object.</dd>
</dl>
<div class="section" id="id3">
<h6>Returns<a class="headerlink" href="#id3" title="Permalink to this headline">¶</a></h6>
<p>An object of the ParametersDict class (a dict of nested dicts, all
of which have key-value pairs whose values can be accessed using typical
dict notation or using dot notation with the keys).</p>
</div>
<div class="section" id="raises">
<h6>Raises<a class="headerlink" href="#raises" title="Permalink to this headline">¶</a></h6>
<dl class="docutils">
<dt>AssertionError</dt>
<dd>If either the Scapes or the Populations parameterized in the parameters
file have not all been given distinct names</dd>
</dl>
</div>
<div class="section" id="id4">
<h6>See Also<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h6>
<p>sim.params.read
sim.params.ParametersDict</p>
</div>
<div class="section" id="id5">
<h6>Examples<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h6>
<p>Read a parameters file called “null_model.py” (located in the current
working directory).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">read_parameters_file</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.params.ParametersDict&#39;&gt;</span>
<span class="go">Model name:                                     GEONOMICS_params_13-10-2018_15:54:03</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="make-model">
<h4><code class="code py python docutils literal"><span class="name"><span class="pre">make_model</span></span></code><a class="headerlink" href="#make-model" title="Permalink to this headline">¶</a></h4>
<p>Create a new Model object.</p>
<p>Use either a ParametersDict object or the path to a valid Geonomics
parameters file (whichever is provided to the ‘parameters’ argument) to
create a new Model object.</p>
<div class="section" id="id6">
<h5>Parameters<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h5>
<dl class="docutils">
<dt>parameters <span class="classifier-delimiter">:</span> <span class="classifier">{ParametersDict, str}, optional</span></dt>
<dd>The parameters to be used to make the Model object.
If <cite>parameters</cite> is a ParametersDict object, the object will be used to
make the Model.
If <cite>parameters</cite> is a string, Geonomics will call
<cite>gnx.read_parameters_file</cite> to make a ParametersDict object, then use
that object to make the Model.
If <cite>parameters</cite> is None, or is not provided, then Geonomics will
attempt to find a single parameters file in the current working
directory with the filename “GEONOMICS_params_&lt;…&gt;.py”, will use that
file to make a ParametersDict object, then will use that object to
make the Model.</dd>
</dl>
<div class="section" id="id7">
<h6>Returns<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h6>
<dl class="docutils">
<dt>out <span class="classifier-delimiter">:</span> <span class="classifier">Model</span></dt>
<dd>An object of the Model class</dd>
</dl>
</div>
<div class="section" id="id8">
<h6>Raises<a class="headerlink" href="#id8" title="Permalink to this headline">¶</a></h6>
<dl class="docutils">
<dt>ValueError</dt>
<dd>If the <cite>parameters</cite> argument was not provided and a single, valid
Geonomics parameters file could not be identified in the current
working directory</dd>
<dt>ValueError</dt>
<dd>If the <cite>parameters</cite> arugment was given a string that does not point
to a valid parameters file</dd>
<dt>ValueError</dt>
<dd>If the ParametersDict provided to the <cite>parameters</cite> argument, or created
from the parameters file being used, cannot be successfully made into a
Model</dd>
</dl>
</div>
<div class="section" id="id9">
<h6>See Also<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h6>
<p>gnx.read_parameters_file
sim.model.Model</p>
</div>
<div class="section" id="id10">
<h6>Examples<a class="headerlink" href="#id10" title="Permalink to this headline">¶</a></h6>
<p>Make a Model from a single, valid “GEONOMICS_params_&lt;…&gt;.py” file that can
be found in the current working directory (such as a file that would be
produced by calling gnx.make_parameters_file without any arguments).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">()</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     GEONOMICS_params_13-10-2018_15:54:03</span>
<span class="go">Scapes:                                         0: &#39;0&#39;</span>
<span class="go">Populations:                                    0: &#39;0&#39;</span>
<span class="go">Number of iterations:                           1</span>
<span class="go">Number of burn-in timesteps (minimum):          30</span>
<span class="go">Number of main timesteps:                       100</span>
<span class="go">Geo-data collected:                             {}</span>
<span class="go">Gen-data collected:                             {}</span>
<span class="go">Stats collected:                                {}</span>
</pre></div>
</div>
<p>Make a Model from a file called ‘null_model.py’, in the current working
directory.</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">gnx</span><span class="o">.</span><span class="n">make_model</span><span class="p">(</span><span class="s1">&#39;null_model.py&#39;</span><span class="p">)</span>
<span class="go">&lt;class &#39;sim.model.Model&#39;&gt;</span>
<span class="go">Model name:                                     null_model</span>
<span class="go">Scapes:                                         0: &#39;tmp&#39;</span>
<span class="go">                                                1: &#39;ppt&#39;</span>
<span class="go">Populations:                                    0: &#39;C. fasciata&#39;</span>
<span class="go">Number of iterations:                           2500</span>
<span class="go">Number of burn-in timesteps (mininum):          100</span>
<span class="go">Number of main timesteps:                       1000</span>
<span class="go">Geo-data collected:                             {csv, geotiff}</span>
<span class="go">Gen-data collected:                             {vcf, fasta}</span>
<span class="go">Stats collected:                                {maf, ld, mean_fit, het, Nt}</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper"><div class="relations">
<h3>Related Topics</h3>
<ul>
  <li><a href="index.html">Documentation overview</a><ul>
  </ul></li>
</ul>
</div>
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="footer">
      &copy;2018, Drew Hart.
      
      |
      Powered by <a href="http://sphinx-doc.org/">Sphinx 1.6.7</a>
      &amp; <a href="https://github.com/bitprophet/alabaster">Alabaster 0.7.8</a>
      
      |
      <a href="_sources/doc.rst.txt"
          rel="nofollow">Page source</a>
    </div>

    

    
  </body>
</html>